---
title: "Host filtering"
author: "James Abbott"
date: "04/10/2021"
output: html_document
---

```{r setup, include=FALSE}
library(knitr)
library(DESeq2)
library(tidyverse)
library(data.table)
library(ggplot2)
library(ggthemes)
library(ggsci)
library(scales)
library(cowplot)
library(phyloseq)
library(microshades)
library(vegan)
library(patchwork)

knitr::opts_chunk$set(echo = TRUE)
opts_knit$set(root.dir = "/cluster/db/jabbott/host_filtering/")
```

A plot of the count of reads discarded under different filtering criteria

```{r filterplot_datapre}
bulk_fastq_counts<-read.table('bulk_fastq_counts.txt',sep="\t",header = TRUE)
bulk_fastq_counts['Group']='Bulk'
bulk_fastq_counts['Replicate']=1:3
elite_fastq_counts<-read.table('elite_fastq_counts.txt',sep="\t",header = TRUE)
elite_fastq_counts['Group']='Elite'
elite_fastq_counts['Replicate']=1:3

merged_fastq_counts=rbind(bulk_fastq_counts,elite_fastq_counts) #%>%
initial_merged_fastq_counts<-merged_fastq_counts %>%
  select(unfiltered,mapped,unmapped,Group,Replicate) %>%
  pivot_longer(!c(Group,Replicate),names_to='Filter')

merged_fastq_counts<-merged_fastq_counts %>%
  pivot_longer(!c(Group,Replicate),names_to='Filter')

merged_fastq_counts$Filter<-factor(merged_fastq_counts$Filter,levels=c('unfiltered','mapped','unmapped','mapq_10','mapq_15','mapq_20','mapq_25','mapq_30'))
```
```{r filter_prop_plot}

initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "unfiltered"] <- "Total"
initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "mapped"] <- "Discarded"
initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "unmapped"] <- "Retained"
initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "unfiltered_prop"] <- "Total %"
initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "mapped_prop"] <- "Discared %"
initial_merged_fastq_counts["Filter"][initial_merged_fastq_counts["Filter"] == "unmapped_prop"] <- "Retained %"

initial_merged_fastq_counts$Filter<-factor(initial_merged_fastq_counts$Filter,
                                           levels=c('Total','Discarded','Retained'))#,'Total %','Mapped %','Unmapped %'))

count.summary <- initial_merged_fastq_counts %>%
  group_by(Group, Filter) %>%
  filter(! grepl('%',Filter)) %>%
  summarise(
    sd = as.integer(sd(value)),
    count = mean(value)
  )

count_plot<-ggplot(count.summary,aes(Filter,count)) +
  geom_col(aes(colour=Group,fill=Group),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  #geom_errorbar(
  #  aes(ymin = count-sd, ymax = count+sd,colour=Group),
  #  position = position_dodge(0.8),width=0.2)+
  geom_point(data=initial_merged_fastq_counts, 
              aes(x=Filter,y=value,colour=Filter),
              size=1,position = position_jitter(width=0.2),colour='darkslategray') + 
  theme_bw()+
  scale_color_startrek()+
  scale_fill_startrek()+
  scale_y_continuous(
    "Mean Read Count",
    labels = scales::label_number_si()
  )+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45, hjust = 1,  vjust = 1))+
  facet_wrap('Group')+
  ggtitle('Filtered Read Counts')
count_plot
ggsave('images/initial_filtered_counts.pdf',plot=count_plot,device='pdf',useDingbats=FALSE,width=20,height=10,units='cm')
```
```{r meanplot}
merged_fastq_counts<-rbind(bulk_fastq_counts,elite_fastq_counts) %>% 
  select(-Replicate) %>%
  pivot_longer(-Group,names_to='Filter')

merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_10"] <- "MapQ 10"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_15"] <- "MapQ 15"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_20"] <- "MapQ 20"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_25"] <- "MapQ 25"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_30"] <- "MapQ 30"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_35"] <- "MapQ 35"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_40"] <- "MapQ 40"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "unfiltered"] <- "Total"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapped"] <- "Discarded"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "unmapped"] <- "Retained"

merged_fastq_counts$Filter<-factor(
    merged_fastq_counts$Filter,
    levels=c('Total','Discarded','Retained','MapQ 10','MapQ 15','MapQ 20','MapQ 25','MapQ 30','MapQ 35','MapQ 40'))

count.summary <- merged_fastq_counts %>%
  group_by(Group, Filter) %>%
  summarise(
    sd = as.integer(sd(value)),
    count = mean(value)
  )

count_plot<-ggplot(count.summary,aes(Filter,count)) +
  geom_col(aes(colour=Group,fill=Group),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=merged_fastq_counts, 
              aes(x=Filter,y=value,colour=Filter),
              size=1,position = position_jitter(width=0.3),colour='darkslategray') + 
  theme_bw()+
  scale_color_startrek()+
  scale_fill_startrek()+
  scale_y_continuous(
    "Read Count",
    labels = scales::label_number_si()
  )+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45, hjust = 1,  vjust = 1))+
  facet_wrap('Group')+
  ggtitle('Filtered Read Counts')
count_plot
ggsave('images/filtered_read_counts.pdf',device='pdf',plot = count_plot, width=20, height=15,units='cm',useDingbats=FALSE)

```
We can also plot the proportion of eukaryotic and host sequence remaining in each set. 

```{r remaininghost}
euk_remains<-read.table('kraken/euk_barley_props.txt',sep="\t",header=TRUE)
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq10"] <- "MapQ 10"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq15"] <- "MapQ 15"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq20"] <- "MapQ 20"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq25"] <- "MapQ 25"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq30"] <- "MapQ 30"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq35"] <- "MapQ 35"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq40"] <- "MapQ 40"
euk_remains["Filter"][euk_remains["Filter"] == "nt_unfiltered"] <- "Total"
euk_remains["Filter"][euk_remains["Filter"] == "nt_morex3_aligned"] <- "Discarded"
euk_remains["Filter"][euk_remains["Filter"] == "nt_morex3_unaligned"] <- "Retained"
initial_euk_remains<-euk_remains %>%
  filter(! grepl('MapQ',Filter))

merged_fastq_counts$Filter<-factor(
    merged_fastq_counts$Filter,
    levels=c('Total','Discarded','Retained','MapQ 10','MapQ 15','MapQ 20','MapQ 25','MapQ 30','MapQ 35','MapQ 40'))

initial_merged_fastq_counts<-merged_fastq_counts %>%
  filter(! grepl('MapQ',Filter))

#initial_merged_fastq_counts$Filter<-factor(
#  initial_merged_fastq_counts,
#  levels=c('Unfiltered','Mapped','Unmapped')
#)

remains.summary <- euk_remains %>%
  group_by(Sample, Filter) %>%
  summarise(
    euk_sd = sd(Eukaryote),
    euk_props = mean(Eukaryote),
    barley_sd = sd(Barley),
    barley_props = mean(Barley)
  )

remains.summary$Filter<-factor(
    remains.summary$Filter,
    levels=c('Total','Discarded','Retained','MapQ 10','MapQ 15','MapQ 20','MapQ 25','MapQ 30','MapQ 35','MapQ 40'))
initial_remains.summary<-remains.summary %>%
  filter(! grepl('MapQ',Filter))

initial_barley_plot<-ggplot(initial_remains.summary,aes(Filter,barley_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=initial_euk_remains, 
              aes(x=Filter,y=Barley),
              size=1,position = position_jitter(width = 0.3),colour='darkslategray') + 
  theme_bw()+
  ylim(0,NA)+
  scale_color_startrek()+
  scale_fill_startrek()+
  ylab('Percentage')+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Barley Proportion')

barley_plot<-ggplot(remains.summary,aes(Filter,barley_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Barley),
              size=1,position = position_jitter(width = 0.3),colour='darkslategray') + 
  theme_bw()+
  ylim(0,NA)+
  scale_color_startrek()+
  scale_fill_startrek()+
  ylab('Percentage')+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Barley Proportion')

barley_plot

initial_euk_plot<-ggplot(initial_remains.summary,aes(Filter,euk_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), 
           width = 0.7) +
  geom_point(data=initial_euk_remains, 
              aes(x=Filter,y=Eukaryote),
              size=1,position = position_dodge2(0.8),colour='darkslategray') + 
  theme_bw()+
  ylab('Percentage')+
  scale_color_startrek()+
  scale_fill_startrek()+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Eukaryote Proportion')


euk_plot<-ggplot(remains.summary,aes(Filter,euk_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), 
           width = 0.7) +
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Eukaryote),
              size=1,position = position_dodge2(0.8),colour='darkslategray') + 
  theme_bw()+
  ylab('Percentage')+
  scale_color_startrek()+
  scale_fill_startrek()+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Eukaryote Proportion')
euk_plot

combined_initial_plot<-initial_euk_plot+initial_barley_plot
combined_initial_plot
ggsave('images/intial_remaining_euk_barley.pdf',plot=combined_initial_plot,device = 'pdf',useDingbats=FALSE,width=20,height=10,units='cm')

combined_plot<-euk_plot+barley_plot
combined_plot
ggsave('images/remaining_euk_barley.pdf',plot=combined_plot,device='pdf',width=20,height=20,units='cm',useDingbats=FALSE)
```

```{r remains_mapq_only}
mapq_remains.summary<-remains.summary %>% filter(grepl('MapQ',Filter))
mapq_euk_remains<-euk_remains %>% filter(grepl('MapQ',Filter))

barley_plot<-ggplot(mapq_remains.summary,aes(Filter,barley_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=mapq_euk_remains, 
              aes(x=Filter,y=Barley),
              size=1,position = position_jitter(width = 0.3),colour='darkslategray') + 
  theme_bw()+
  ylim(0,NA)+
  scale_color_startrek()+
  scale_fill_startrek()+
  ylab('Percentage')+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Barley Proportion')

barley_plot

euk_plot<-ggplot(mapq_remains.summary,aes(Filter,euk_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), 
           width = 0.7) +
  geom_point(data=mapq_euk_remains, 
              aes(x=Filter,y=Eukaryote),
              size=1,position = position_dodge2(0.8),colour='darkslategray') + 
  theme_bw()+
  ylab('Percentage')+
  scale_color_startrek()+
  scale_fill_startrek()+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Eukaryote Proportion')
euk_plot
combined_plot<-euk_plot+barley_plot
combined_plot
ggsave('images/remaining_euk_barley_mapq.pdf',plot=combined_plot,device='pdf',width=20,height=15,units='cm',useDingbats=FALSE)
```

```{r kraken_dataprep}
get_counts<-function(headname) {
  name_parts<-(strsplit(headname,'\\.'))[[1]]
  
  if (name_parts[[1]]=='pre') {
    dir='nt_unfiltered'
  } else {
    dir='nt_morex3_unaligned'
  }
  
  path=file.path('kraken',dir,paste0(name_parts[[2]],'.report.txt'))
  
  df<-read.csv(path,sep="\t",col.names=c('prop',headname,'count','rank','taxid','taxa')) %>% 
    filter(rank=='G') %>%
    select(taxa,taxid,all_of(headname))
  
  df$taxa<-trimws(df$taxa,which='left')
    
  return(df)
}

# just interested in bulk and elite here, but will pick them pre and post filtering
bulk_samples=c('pre.2000','pre.2001','pre.2002','post.2000','post.2001','post.2002')
elite_samples=c('pre.2023','pre.2024','pre.2025','post.2023','post.2024','post.2025')

bulk_df<-lapply(bulk_samples, get_counts) %>%
  reduce(left_join,by='taxa') %>%
  replace(is.na(.),0) %>%
  column_to_rownames('taxa') %>%
  select(taxid.x,unlist(bulk_samples))

elite_df<-lapply(elite_samples,get_counts) %>%
  reduce(left_join,by='taxa') %>%
  replace(is.na(.),0) %>%
  column_to_rownames('taxa') %>%
  select(taxid.x,unlist(elite_samples))

phylum_scale_4 = scale_fill_manual(name='Phylum',
                            values=c( "#9ecae1", "#a1d99b", "#41ab5d","#238b45" ))
phylum_scale_8 = scale_fill_manual(name='Phylum',
                            values=c("#B78560", "#bcbddc", "#9e9ac8", "#6a51a3", 
                                     "#9ecae1",  "#a1d99b", "#41ab5d","#238b45" ))

class_scale = scale_fill_manual(name='Class',
                                 values=c("#CAA995", "#B78560", "#9E5C00", "#7D3200",
                                          "#bcbddc", "#9e9ac8", "#807dba", "#6a51a3",
                                          "#c6dbef", "#9ecae1", "#6baed6", "#4292c6",
                                          "#fec44f", "#fdae6b", "#fe9929", "#ff7f00",
                                          "#a1d99b", "#74c476", "#41ab5d", "#238b45"
                                # values=c("#dadaeb", "#bcbddc", "#9e9ac8", "#807dba", "#6a51a3",
                                #          "#eff3ff", "#c6dbef", "#9ecae1", ""#807dba", "#6a51a3",
                                #          "#feeda0", "#fec44f", "#fdae6b", "#807dba", "#6a51a3",
                                #          "#c7e9c0", "#a1d99b", "#74c476", "#41ab5d", "#238b45",
                                #          "#D8C7BE", "#CAA995", "#B78560", "#9E5C00", "#7D3200"
                                         ))
```

```{r DeSeq2}
bulk_counts<-bulk_df %>% 
  select(-taxid.x) %>% 
  as.matrix()
condition=factor(c(rep('pre',3),c(rep('post',3))))
coldata<-data.frame(row.names=colnames(bulk_counts),condition)

dds<-DESeqDataSetFromMatrix(countData=bulk_counts,colData=coldata,design= ~ condition)
keep <- rowSums(counts(dds)) >= 50
dds <- dds[keep,]
dds<-DESeq(dds)
rld <- rlogTransformation(dds)
res<-results(dds,alpha=0.05, contrast=c("condition",'pre','post'))
summary(res)

bulk_df<-rownames_to_column(bulk_df,var='taxa')
bulk_res<-as.data.frame(res)
bulk_res <- bulk_res[order(bulk_res$log2FoldChange), ] %>%
  filter(padj <= 0.05) %>%
  rownames_to_column(var='taxa') %>%
  left_join(bulk_df,by='taxa')

elite_counts<-elite_df %>%
  select(-taxid.x) %>%
  as.matrix()
  
dds<-DESeqDataSetFromMatrix(countData=elite_counts,colData=coldata,design= ~ condition)
dds<-DESeq(dds)

rld <- rlogTransformation(dds)
res<-results(dds,alpha=0.05, contrast=c("condition",'pre','post'))
summary(res)

elite_df<-rownames_to_column(elite_df,var='taxa')
elite_res<-as.data.frame(res)
elite_res <- elite_res[order(elite_res$log2FoldChange), ] %>%
  filter(padj <= 0.05) %>%
  rownames_to_column(var='taxa') %>%
  left_join(elite_df,by='taxa')
```
```{r output}
dir.create('deseq2', showWarnings = FALSE)

bulk_res<-bulk_res %>%
  relocate('taxid.x',.after = 'taxa') %>%
  rename('taxid'='taxid.x')

elite_res<-elite_res %>%
  relocate('taxid.x',.after = 'taxa') %>%
  rename('taxid'='taxid.x')

write.table(bulk_res,file.path('deseq2','bulk.txt'),sep="\t",quote=FALSE,row.names = FALSE)
write.table(elite_res,file.path('deseq2','elite.txt'),sep="\t",quote=FALSE,row.names = FALSE)
```

## Phyloseq analysis of kraken outputs

Phyloseq reports have been converted into a biom file using kraken-biom, which will hopefully load into phyloseq ok...

```{r phyloseq_prep}
elite_samples_df<-read.csv('elite_map.txt',header = TRUE,sep="\t")
bulk_samples_df<-read.csv('bulk_map.txt',header = TRUE,sep="\t")

initial_elite_samples_df<-elite_samples_df %>%
  filter(! grepl('mapq', SampleID))
initial_bulk_samples_df<-bulk_samples_df %>%
  filter(! grepl('mapq', SampleID))

mapq_elite_samples_df<-elite_samples_df %>%
  filter(! grepl("MapQ[0-9]5",Status))
mapq_bulk_samples_df<-bulk_samples_df %>%
  filter(! grepl("MapQ[0-9]5",Status))

row.names(initial_elite_samples_df)<-initial_elite_samples_df$SampleID
row.names(initial_bulk_samples_df)<-initial_bulk_samples_df$SampleID
row.names(elite_samples_df)<-elite_samples_df$SampleID
row.names(bulk_samples_df)<-bulk_samples_df$SampleID
row.names(mapq_elite_samples_df)<-mapq_elite_samples_df$SampleID
row.names(mapq_bulk_samples_df)<-mapq_bulk_samples_df$SampleID

initial_elite_samples<-initial_elite_samples_df$SampleID
initial_bulk_samples<-initial_bulk_samples_df$SampleID
elite_samples<-elite_samples_df$SampleID
bulk_samples<-bulk_samples_df$SampleID
mapq_elite_samples<-mapq_elite_samples_df$SampleID
mapq_bulk_samples<-mapq_bulk_samples_df$SampleID

initial_elite_SAM = sample_data(initial_elite_samples_df, errorIfNULL = T)
initial_bulk_SAM = sample_data(initial_bulk_samples_df, errorIfNULL = T)
elite_SAM = sample_data(elite_samples_df, errorIfNULL = T)
bulk_SAM = sample_data(bulk_samples_df, errorIfNULL = T)
mapq_elite_SAM = sample_data(mapq_elite_samples_df, errorIfNULL = T)
mapq_bulk_SAM = sample_data(mapq_bulk_samples_df, errorIfNULL = T)

ps<-import_biom('kraken/biom/table.biom.gz')
colnames(tax_table(ps))<-c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
# clean up taxtable entries to remove k__, c__ etc prefixes
tax_table(ps)[, colnames(tax_table(ps))] <- gsub(tax_table(ps)[, colnames(tax_table(ps))], pattern = "[a-z]__", replacement = "")

bact_ps<-subset_taxa(ps,Kingdom=='Bacteria')

initial_elite_bact_ps<-prune_samples(initial_elite_samples,bact_ps)
initial_bulk_bact_ps<-prune_samples(initial_bulk_samples,bact_ps)
elite_bact_ps<-prune_samples(elite_samples,bact_ps)
bulk_bact_ps<-prune_samples(bulk_samples,bact_ps)
mapq_elite_bact_ps<-prune_samples(mapq_elite_samples,bact_ps)
mapq_bulk_bact_ps<-prune_samples(mapq_bulk_samples,bact_ps)


initial_elite_bact_ps<-phyloseq(otu_table(initial_elite_bact_ps@otu_table),tax_table(initial_elite_bact_ps@tax_table),initial_elite_SAM)
initial_bulk_bact_ps<-phyloseq(otu_table(initial_bulk_bact_ps@otu_table),tax_table(initial_bulk_bact_ps@tax_table),initial_bulk_SAM)
elite_bact_ps<-phyloseq(otu_table(elite_bact_ps@otu_table),tax_table(elite_bact_ps@tax_table),elite_SAM)
bulk_bact_ps<-phyloseq(otu_table(bulk_bact_ps@otu_table),tax_table(bulk_bact_ps@tax_table),bulk_SAM)
mapq_elite_bact_ps<-phyloseq(otu_table(mapq_elite_bact_ps@otu_table),tax_table(mapq_elite_bact_ps@tax_table),mapq_elite_SAM)
mapq_bulk_bact_ps<-phyloseq(otu_table(mapq_bulk_bact_ps@otu_table),tax_table(mapq_bulk_bact_ps@tax_table),mapq_bulk_SAM)
```

A function to return a factor of taxa ordered by abundance...
```{r get_tax_order}
get_tax_count<-function(df,taxrank,taxa) {
   taxa_df<-df %>% filter(get(taxrank)==taxa) 
   return(c(taxa,sum(taxa_df$Abundance)))
}

get_tax_order<-function(df,taxrank) {
  levs=levels(factor(df[[taxrank]]))
  taxa_abundance<-sapply(levs, function(x) {get_tax_count(df,taxrank,x)})
  taxa_abundance<-data.frame(t(taxa_abundance)) 
  taxa_abundance$X2<-as.numeric(taxa_abundance$X2)
  taxa_abundance<-taxa_abundance %>% arrange(X2)
  
  return(factor(taxa_abundance$X1))
}

```



```{r phylum_merged_relative_plot}
## And merging replicates...For relative values the merge should be done on 
# absolute values then rescaled
glommed<-tax_glom(elite_bact_ps,taxrank='Phylum')
derep<-merge_samples(glommed,'Status')
phy<-transform_sample_counts(derep,function(x) x/sum(x))
dat<-data.table(psmelt(phy))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Change name to remainder of Phylum less than arbitrary cutoff to match 1%...
dat[(median <= 0.007), Phylum := "Others"]

dat<-dat %>% mutate(Sample = case_when(
  Sample=='Unfiltered' ~ 'Total',
  Sample=='MorexV3 aligned' ~ 'Discarded',
  Sample=='MorexV3 unaligned' ~ 'Retained',
  Sample=='MapQ10' ~ 'MapQ10',
  Sample=='MapQ15' ~ 'MapQ15',
  Sample=='MapQ20' ~ 'MapQ20',
  Sample=='MapQ25' ~ 'MapQ25',
  Sample=='MapQ30' ~ 'MapQ30'
))


phyla=get_tax_order(dat,'Phylum')
#dat$Filtered=factor(dat$Filtered,levels=c('Unfiltered','Filtered'))
dat$Sample=factor(dat$Sample,levels=c('Total','Discarded','Retained','MapQ10','MapQ15','MapQ20','MapQ25','MapQ30'))

initial_dat<-dat %>%
  filter(! grepl('MapQ',Sample))

initial_elite_relative_merged_plot <- ggplot(data=initial_dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(axis.ticks.x = element_blank(), 
        axis.title.x = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        legend.position = 'bottom',
        legend.text=element_text(size=8),
        legend.key.size = unit(0.4, 'cm'),
        legend.title=element_blank())+
  phylum_scale_8+
  ylab('Relative Abundance')+
  ggtitle('Elite Phylum Relative Abundance')



initial_elite_relative_merged_plot
ggsave('images/initial_elite_relative_merged_phylum.pdf',plot=initial_elite_relative_merged_plot,device='pdf',
       useDingbats=FALSE,width=13,height=10,units='cm')

elite_relative_merged_plot <- ggplot(data=dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(axis.ticks.x = element_blank(), 
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        strip.background = element_rect(
          color="black", fill="white", size=0.5, linetype="solid"
        ),
        strip.text = element_text(size=7),
        legend.position = 'bottom',
        legend.text=element_text(size=8),
        legend.key.size = unit(0.4, 'cm'),
        legend.title=element_text(size=10))+
  phylum_scale_8+
  ylab('Relative Abundance')+
  ggtitle('Elite Phylum Relative Abundance')
elite_relative_merged_plot
```
```{r phylum_merged_rep_plot}
## And merging replicates...
derep<-merge_samples(glommed,'Status')
dat<-data.table(psmelt(derep))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Arbitrary cutoff for median to match number of phyla in previous plots
dat[(median <= 140000), Phylum := "Others"]
 
dat<-dat %>% mutate(Sample = case_when(
  Sample=='Unfiltered' ~ 'Total',
  Sample=='MorexV3 aligned' ~ 'Discarded',
  Sample=='MorexV3 unaligned' ~ 'Retained'
))
 
initial_dat<-dat %>%
  filter(! grepl ('MapQ',Sample))

phyla=get_tax_order(dat,'Phylum')
dat$Sample=factor(dat$Sample,levels=c('Total','Discarded','Retained'))
 
initial_elite_abs_merged_plot <- ggplot(data=initial_dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(legend.position = "bottom", 
        axis.ticks.x = element_blank(), 
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        strip.background = element_rect(
          color="black", fill="white", size=0.5, linetype="solid"
        ),
        strip.text = element_text(size=7)) +
  phylum_scale_8+
  xlab('Condition')+
  ylab('Absolute Abundance')
#elite_abs_merged_plot
```
```{r merge_phylum_plots}
phylum_plot<-initial_elite_abs_merged_plot+initial_elite_relative_merged_plot & theme(legend.position = "bottom")
phylum_plot<-phylum_plot+plot_layout(guides = "collect")
phylum_plot
```

## Bulk Phylum Plots

```{r bulk_phylum_abundance_plots}
glommed<-tax_glom(bulk_bact_ps,taxrank='Phylum')
dat<-data.table(psmelt(glommed))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Change name to remainder of Phylum less than arbitrary cutoff to match 1%...
dat[(median <= 60000), Phylum := "Others"]

dat<-dat %>% mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Unfiltered',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained'
))

phyla=get_tax_order(dat,'Phylum')
dat$Filtered=factor(dat$Filtered,levels=c('Unfiltered','Filtered'))
dat$Status=factor(dat$Status,levels=c('Unfiltered','Discarded','Retained'))

bulk_abs_replicate_plot <- ggplot(data=dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
        axis.ticks.x = element_blank(), 
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        strip.background = element_rect(
          color="black", fill="white", size=0.5, linetype="solid"
        ),
        strip.text = element_text(size=7)) +
  phylum_scale_8 +
  xlab('')+
  ylab('Absolute Abundance')+
  facet_grid(~ Status,scales="free",space="free")
bulk_abs_replicate_plot
```


```{r bulk_phylum_merged_rep_plot}
## And merging replicates...
derep<-merge_samples(glommed,'Status')
dat<-data.table(psmelt(derep))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Arbitrary cutoff for median to match number of phyla in previous plots
#dat[(median <= 58000), Phylum := "Others"]
dat[(median <= 200000), Phylum := "Others"]

dat<-dat %>% mutate(Sample = case_when(
  Sample=='Unfiltered' ~ 'Unfiltered',
  Sample=='MorexV3 aligned' ~ 'Discarded',
  Sample=='MorexV3 unaligned' ~ 'Retained'
))

phyla=get_tax_order(dat,'Phylum')
dat$Filtered=factor(dat$Filtered,levels=c('Unfiltered','Filtered'))
dat$Sample=factor(dat$Sample,levels=c('Unfiltered','Discarded','Retained'))

bulk_abs_merged_plot <- ggplot(data=dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(legend.position = "none", 
        axis.ticks.x = element_blank(), 
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        strip.background = element_rect(
          color="black", fill="white", size=0.5, linetype="solid"
        ),
        strip.text = element_text(size=7)) +
  phylum_scale_4+
  xlab('Condition')+
  ylab('Absolute Abundance')
bulk_abs_merged_plot
```
```{r phylum_relative_replicate_plot}
phy<-transform_sample_counts(bulk_bact_ps,function(x) x/sum(x))
glommed<-tax_glom(phy,taxrank='Phylum')
dat<-data.table(psmelt(glommed))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE),  by = "Phylum"]
# Change name to remainder of Phylum less than 1%
dat[(median <= 0.01), Phylum := "Others"]

dat<-dat %>% mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained'
))

initial_dat<-dat %>%
  filter(! grepl('MapQ',Sample))
dat$Filtered=factor(dat$Filtered,levels=c('Unfiltered','Filtered'))
dat$Status=factor(dat$Status,levels=c('Unfiltered','Discarded','Retained'))
phyla=get_tax_order(dat,'Phylum')

bulk_relative_replicate_plot <- ggplot(data=initial_dat, aes(x=Sample, y=Abundance, fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + theme(legend.position='bottom',
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     plot.margin = unit(c(0.5, 0.5, 0.0, 0.5), "cm"),
                     strip.background = element_rect(
                       color="black", fill="white", size=0.5, linetype="solid"
                     ),
                     strip.text = element_text(size=7),
                     legend.text=element_text(size=8),
                     legend.key.size = unit(0.4, 'cm'),
                     legend.title=element_text(size=10)
                ) + 
  phylum_scale_8 +
  ylab('Relative Abundance')

legend<-get_legend(bulk_relative_replicate_plot)
bulk_relative_replicate_plot<-bulk_relative_replicate_plot + theme(legend.position = 'none')
p_title=ggplot()+
  ggtitle('Effect of reference filtering on Bulk samples: Separate replicates')+
  theme(plot.margin = unit(c(0.5, 0.5, 0.0, 0.5), "cm"))
bulk_relative_replicate_plot
bulk_replicate_plots<-plot_grid(plot_grid(p_title,ncol=1),
             plot_grid(bulk_abs_replicate_plot,bulk_relative_replicate_plot,ncol=2),
             plot_grid(legend,ncol=1),
             nrow=3,greedy = FALSE,rel_heights = c(0.1,1,0.1))+
  theme(plot.background = element_rect(fill='white',colour='white'))
ggsave('images/bulk_phylum_abundance.png',plot = bulk_replicate_plots, device='png',width=20,height=10,units='cm')
bulk_replicate_plots
```
```{r bulkl_phylum_merged_relative_plot}
## And merging replicates...For relative values the merge should be done on 
# absolute values then rescaled
glommed<-tax_glom(bulk_bact_ps,taxrank='Phylum')
derep<-merge_samples(glommed,'Status')
phy<-transform_sample_counts(derep,function(x) x/sum(x))
dat<-data.table(psmelt(phy))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Change name to remainder of Phylum less than arbitrary cutoff to match 1%...
dat[(median <= 0.001), Phylum := "Others"]

initial_dat<-dat %>%
  filter(! grepl('MapQ',Sample))

dat<-dat %>% mutate(Sample = case_when(
  Sample=='Unfiltered' ~ 'Total',
  Sample=='MorexV3 aligned' ~ 'Discarded',
  Sample=='MorexV3 unaligned' ~ 'Retained'
))

phyla=get_tax_order(dat,'Phylum')
dat$Sample=factor(dat$Sample,levels=c('Total','Discarded','Retained'))

bulk_relative_merged_plot <- ggplot(data=initial_dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(axis.ticks.x = element_blank(), 
        axis.title.x = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        legend.position = 'bottom',
        legend.text=element_text(size=8),
        legend.key.size = unit(0.4, 'cm'),
        legend.title=element_blank())+
  phylum_scale_8+
  ylab('Relative Abundance')+
  ggtitle('Bulk Phylum Relative Abundance')

bulk_relative_merged_plot
ggsave('images/initial_bulk_relative_merged_phylum.pdf',plot=bulk_relative_merged_plot,device='pdf',
       useDingbats=FALSE,width=13,height=10,units='cm')
```
```{r merge_bulk_relative_plots}
legend<-get_legend(bulk_relative_merged_plot)
bulk_relative_merged_plot<-bulk_relative_merged_plot + theme(legend.position = 'none')

p_title=ggplot()+
  ggtitle('Effect of reference filtering on Bulk samples: Merged replicates')+
  theme(plot.margin = unit(c(0.5, 0.5, 0.0, 0.5), "cm"))

merged_plots<-plot_grid(plot_grid(p_title,ncol=1),
             plot_grid(bulk_abs_merged_plot,bulk_relative_merged_plot,ncol=2),
             plot_grid(legend,ncol=1),
             nrow=3,greedy = FALSE,rel_heights = c(0.1,1,0.2))+
  theme(plot.background = element_rect(fill='white',colour='white'))
ggsave('images/bulk_phylum_abundance_merged.png',plot = merged_plots, device='png',width=20,height=10,units='cm')
merged_plots
```

We can get an idea of whether there has been a material impact on the community from a beta-diversity comparison...

```{r bray_distance}
 initial_elite_samples_df<-initial_elite_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained'
))
initial_bulk_samples_df<-initial_bulk_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained'
))

bulk_samples_df<-bulk_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

elite_samples_df<-elite_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

mapq_bulk_samples_df<-mapq_bulk_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

mapq_elite_samples_df<-mapq_elite_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

initial_elite_SAM = sample_data(initial_elite_samples_df, errorIfNULL = T)
initial_bulk_SAM = sample_data(initial_bulk_samples_df, errorIfNULL = T)
elite_SAM = sample_data(elite_samples_df, errorIfNULL = T)
bulk_SAM = sample_data(bulk_samples_df, errorIfNULL = T)
mapq_elite_SAM = sample_data(mapq_elite_samples_df, errorIfNULL = T)
mapq_bulk_SAM = sample_data(mapq_bulk_samples_df, errorIfNULL = T)

initial_elite_bact_ps<-phyloseq(otu_table=initial_elite_bact_ps@otu_table,
                                tax_table=initial_elite_bact_ps@tax_table,
                                sam_data=initial_elite_SAM)

initial_bulk_bact_ps<-phyloseq(otu_table=initial_bulk_bact_ps@otu_table,
                                tax_table=initial_bulk_bact_ps@tax_table,
                                sam_data=initial_bulk_SAM)

elite_bact_ps<-phyloseq(otu_table=elite_bact_ps@otu_table,
                        tax_table=elite_bact_ps@tax_table,
                        sam_data=elite_SAM)

bulk_bact_ps<-phyloseq(otu_table=bulk_bact_ps@otu_table,
                        tax_table=bulk_bact_ps@tax_table,
                        sam_data=bulk_SAM)

mapq_elite_bact_ps<-phyloseq(otu_table=mapq_elite_bact_ps@otu_table,
                        tax_table=mapq_elite_bact_ps@tax_table,
                        sam_data=mapq_elite_SAM)

mapq_bulk_bact_ps<-phyloseq(otu_table=mapq_bulk_bact_ps@otu_table,
                        tax_table=mapq_bulk_bact_ps@tax_table,
                        sam_data=mapq_bulk_SAM)
distmat<-phyloseq::distance(initial_bulk_bact_ps,method='bray')
mds<-ordinate(initial_bulk_bact_ps,"MDS",distance=distmat)
dist_plot1<-plot_ordination(initial_bulk_bact_ps,mds,color = 'Status',
                           title = 'Bulk Samples Bray-Curtis PCoA')+
    theme_bw()+
    scale_color_colorblind()+
    geom_point(size=4)

dist_plot1$data$Status=factor(dist_plot1$data$Status,
                              levels=c('Total','Discarded','Retained'))

distmat<-phyloseq::distance(initial_elite_bact_ps,method='bray')
mds<-ordinate(initial_elite_bact_ps,"MDS",distance=distmat)
dist_plot2<-plot_ordination(initial_elite_bact_ps,mds,color = 'Status',
                           title = 'Elite Samples Bray-Curtis PCoA')+
  theme_bw()+
  scale_color_colorblind()+
  geom_point(size=4)

dist_plot2$data$Status=factor(dist_plot2$data$Status,
                              levels=c('Total','Discarded','Retained'))

dist_plot<-dist_plot1+dist_plot2 & theme(legend.position = "bottom",legend.title=element_blank())
dist_plot<-dist_plot+plot_layout(guides = "collect")
ggsave('images/bray.pdf',plot=dist_plot, device='pdf',width=30,height=15,units='cm',useDingbats=FALSE)
```

```{r adonis1}
distmat<-phyloseq::distance(initial_bulk_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(initial_bulk_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res

distmat<-phyloseq::distance(initial_elite_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(initial_elite_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res
```

```{r bray2}
bray_samples<-c('Total','Discarded','Retained','MapQ10','MapQ20','MapQ30','MapQ40')

distmat<-phyloseq::distance(mapq_bulk_bact_ps,method='bray')
mds<-ordinate(bulk_bact_ps,"MDS",distance=distmat)
dist_plot1<-plot_ordination(bulk_bact_ps,mds,color = 'Status',
                           title = 'Bulk Samples Bray-Curtis PCoA')+
  theme_bw()+
  scale_color_colorblind()+
  geom_point(size=4)

dist_plot1$data$Status=factor(dist_plot1$data$Status,
                              levels=bray_samples)

distmat<-phyloseq::distance(mapq_elite_bact_ps,method='bray')
mds<-ordinate(elite_bact_ps,"MDS",distance=distmat)
dist_plot2<-plot_ordination(elite_bact_ps,mds,color = 'Status',
                           title = 'Elite Samples Bray-Curtis PCoA')+
  theme_bw()+
  scale_color_colorblind()+
  geom_point(size=4)

dist_plot2$data$Status=factor(dist_plot2$data$Status,
                              levels=bray_samples)

dist_plot<-dist_plot1+dist_plot2 & theme(legend.position = "bottom",legend.title=element_blank())
dist_plot<-dist_plot+plot_layout(guides = "collect")
ggsave('images/bray2.pdf',plot=dist_plot, device='pdf',width=30,height=15,units='cm',useDingbats=FALSE)
```
```{r adonis2}
distmat<-phyloseq::distance(mapq_bulk_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(mapq_bulk_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res

distmat<-phyloseq::distance(mapq_elite_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(mapq_elite_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res
```

And a fresh plot of counts with the filtered data...

```{r filterplot_datapre}
bulk_fastq_counts<-read.table('bulk_fastq_counts_q30_paired.txt',sep="\t",header = TRUE)
bulk_fastq_counts['Group']='Bulk'
bulk_fastq_counts['Replicate']=1:3
elite_fastq_counts<-read.table('elite_fastq_counts_q30_paired.txt',sep="\t",header = TRUE)
elite_fastq_counts['Group']='Elite'
elite_fastq_counts['Replicate']=1:3

merged_fastq_counts=rbind(bulk_fastq_counts,elite_fastq_counts) %>%
  pivot_longer(!c(Group,Replicate),names_to='Filter')

merged_fastq_counts$Filter<-factor(merged_fastq_counts$Filter,levels=c('unfiltered','mapped','unmapped','mapq_10','mapq_15','mapq_20','mapq_25','mapq_30'))
```
```{r meanplot}
merged_fastq_counts<-rbind(bulk_fastq_counts,elite_fastq_counts) %>% 
  select(-Replicate) %>%
  pivot_longer(-Group,names_to='Filter')


merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_30_paired"] <- "MapQ 30 Paired"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapq_30"] <- "MapQ 30"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "unfiltered"] <- "Total"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "mapped"] <- "Mapped"
merged_fastq_counts["Filter"][merged_fastq_counts["Filter"] == "unmapped"] <- "Unmapped"

merged_fastq_counts$Filter<-factor(
    merged_fastq_counts$Filter,
    levels=c('Total','Mapped','Unmapped','MapQ 30', 'MapQ 30 Paired'))

count.summary <- merged_fastq_counts %>%
  group_by(Group, Filter) %>%
  summarise(
    sd = as.integer(sd(value)),
    count = mean(value)
  )

count_plot<-ggplot(count.summary,aes(Filter,count)) +
  geom_col(aes(colour=Group,fill=Group),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_errorbar(
    aes(ymin = count-sd, ymax = count+sd,colour=Group),
    position = position_dodge(0.8),width=0.2)+
  geom_point(data=merged_fastq_counts, 
              aes(x=Filter,y=value,colour=Filter),
              size=1,position = position_jitter(width=0.3),colour='darkslategray') + 
  theme_bw()+
  scale_color_startrek()+
  scale_fill_startrek()+
  scale_y_continuous(
    "Read Count",
    labels = scales::label_number_si()
  )+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45, hjust = 1,  vjust = 1))+
  facet_wrap('Group')+
  ggtitle('Filtered Read Counts')
count_plot
ggsave('images/filtered_read_counts_mapq30_paired.pdf',device='pdf',plot = count_plot, width=20, height=15,units='cm',useDingbats=FALSE)

```

Replot proportion of remaining eukaryotic/barley sequences following filtering

```{r remains_mapq_only}
euk_remains<-read.table('kraken/euk_barley_props_filtered.txt',sep="\t",header=TRUE)
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq10"] <- "MapQ 10"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq15"] <- "MapQ 15"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq20"] <- "MapQ 20"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq25"] <- "MapQ 25"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq30"] <- "MapQ 30"
euk_remains["Filter"][euk_remains["Filter"] == "nt_mapq30_paired"] <- "MapQ 30 Paired"
euk_remains<-euk_remains %>% filter(grepl('MapQ',Filter))
  
remains.summary <- euk_remains %>%
  filter((grepl('MapQ',Filter))) %>%
  group_by(Sample, Filter) %>%
  summarise(
    euk_sd = sd(Eukaryote),
    euk_props = mean(Eukaryote),
    barley_sd = sd(Barley),
    barley_props = mean(Barley)
  )

barley_plot<-ggplot(remains.summary,aes(Filter,barley_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_errorbar(
    aes(ymin = barley_props-barley_sd, ymax = barley_props+barley_sd,colour=Sample),
    position = position_dodge(0.8),width=0.2)+
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Barley),
              size=1,position = position_jitter(width = 0.3),colour='darkslategray') + 
  theme_bw()+
  ylim(0,NA)+
  scale_color_startrek()+
  scale_fill_startrek()+
  ylab('Percentage')+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Barley Proportion')

barley_plot

euk_plot<-ggplot(remains.summary,aes(Filter,euk_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), 
           width = 0.7) +
  geom_errorbar(
    aes(ymin = euk_props-euk_sd, ymax = euk_props+euk_sd,colour=Sample),
    position = position_dodge(0.8),width=0.2)+
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Eukaryote),
              size=1,position = position_dodge2(0.8),colour='darkslategray') + 
  theme_bw()+
  ylab('Percentage')+
  scale_color_startrek()+
  scale_fill_startrek()+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Eukaryote Proportion')
euk_plot
combined_plot<-plot_grid(euk_plot,barley_plot,ncol=1)
combined_plot
ggsave('images/remaining_euk_barley_mapq_paired.pdf',plot=combined_plot,device='pdf',width=20,height=20,units='cm',useDingbats=FALSE)
```

```{r phyloseq_prep}
elite_samples_df<-read.csv('elite_map_paired.txt',header = TRUE,sep="\t")
bulk_samples_df<-read.csv('bulk_map_paired.txt',header = TRUE,sep="\t")

row.names(elite_samples_df)<-elite_samples_df$SampleID
row.names(bulk_samples_df)<-bulk_samples_df$SampleID

bulk_samples_df<-bulk_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))


elite_samples_df<-elite_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

elite_samples<-row.names(elite_samples_df)
bulk_samples<-row.names(bulk_samples_df)

elite_SAM = sample_data(elite_samples_df, errorIfNULL = T)
bulk_SAM = sample_data(bulk_samples_df, errorIfNULL = T)

ps<-import_biom('kraken/biom/table_mapq30_paired.biom.gz')
colnames(tax_table(ps))<-c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
# clean up taxtable entries to remove k__, c__ etc prefixes
tax_table(ps)[, colnames(tax_table(ps))] <- gsub(tax_table(ps)[, colnames(tax_table(ps))], pattern = "[a-z]__", replacement = "")

bact_ps<-subset_taxa(ps,Kingdom=='Bacteria')

elite_bact_ps<-prune_samples(elite_samples,bact_ps)
bulk_bact_ps<-prune_samples(bulk_samples,bact_ps)

elite_bact_ps<-phyloseq(otu_table(elite_bact_ps),tax_table(elite_bact_ps),elite_SAM)
bulk_bact_ps<-phyloseq(otu_table(bulk_bact_ps),tax_table(bulk_bact_ps),bulk_SAM)
```

```{r bray_distance_3}
distmat<-phyloseq::distance(elite_bact_ps,method='bray')
mds<-ordinate(elite_bact_ps,"MDS",distance=distmat)

elite_dist_plot<-plot_ordination(elite_bact_ps,mds,color = 'Status',title = 'Elite Samples Bray-Curtis PCoA')+
  theme_bw()+
  theme(legend.position = 'bottom')+
  scale_color_colorblind()+
  geom_point(size=4)

elite_dist_plot$data$Status=factor(elite_dist_plot$data$Status,
                                   levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired'))
elite_dist_plot<-elite_dist_plot+theme(legend.position='none')

distmat<-phyloseq::distance(bulk_bact_ps,method='bray')
mds<-ordinate(bulk_bact_ps,"MDS",distance=distmat)

bulk_dist_plot<-plot_ordination(bulk_bact_ps,mds,color = 'Status',title = 'Bulk Bray-Curtis PCoA')+
  theme_bw()+
  theme(legend.position = 'bottom')+
  scale_color_colorblind()+
  geom_point(size=4)

bulk_dist_plot$data$Status=factor(bulk_dist_plot$data$Status,
                                   levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired'))
dist_plot=bulk_dist_plot+elite_dist_plot & theme(legend.position='bottom',legend.title = element_blank())
dist_plot<-dist_plot+plot_layout(guides = "collect")
ggsave('images/bray3.pdf',plot=dist_plot, device='pdf',width=30,height=15,units='cm',useDingbats=FALSE)
```

```{r adonis}
design<-as.data.frame(as.matrix(sample_data(elite_bact_ps)))
adonis(distmat~Status, data= design, permutations = 10000)
```

```{r q30_bulk_phylum_merged_relative_plot}
## And merging replicates...For relative values the merge should be done on 
# absolute values then rescaled
glommed<-tax_glom(bulk_bact_ps,taxrank='Phylum')
derep<-merge_samples(glommed,'Status')
phy<-transform_sample_counts(derep,function(x) x/sum(x))
dat<-data.table(psmelt(phy))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Change name to remainder of Phylum less than arbitrary cutoff to match 1%...
dat[(median <= 0.0022), Phylum := "Others"]

dat$Sample=factor(dat$Sample,levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired'))

phyla=get_tax_order(dat,'Phylum')

bulk_relative_merged_plot <- ggplot(data=dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(axis.ticks.x = element_blank(), 
        axis.title.x = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        legend.position = 'bottom',
        legend.text=element_text(size=8),
        legend.key.size = unit(0.4, 'cm'),
        legend.title=element_blank())+
  phylum_scale_8+
  ylab('Relative Abundance')+
  ggtitle('Bulk Phylum Relative Abundance')

bulk_relative_merged_plot
ggsave('images/q30_bulk_relative_merged_phylum.pdf',plot=bulk_relative_merged_plot,device='pdf',
       useDingbats=FALSE,width=13,height=10,units='cm')

glommed<-tax_glom(elite_bact_ps,taxrank='Phylum')
derep<-merge_samples(glommed,'Status')
phy<-transform_sample_counts(derep,function(x) x/sum(x))
dat<-data.table(psmelt(phy))
dat$Phylum<-as.character(dat$Phylum)
dat[, median := median(Abundance, na.rm = TRUE), by = "Phylum"]
# Change name to remainder of Phylum less than arbitrary cutoff to match 1%...
dat[(median <= 0.007), Phylum := "Others"]
dat$Sample=factor(dat$Sample,levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired'))

phyla=get_tax_order(dat,'Phylum')

elite_relative_merged_plot <- ggplot(data=dat, aes(x=Sample, y=Abundance,fill=factor(Phylum,levels=phyla))) + 
  geom_bar(aes(), stat="identity", position="stack") +
  theme_bw() + 
  theme(axis.ticks.x = element_blank(), 
        axis.title.x = element_blank(),
        plot.margin = unit(c(0.5, 0.5, 0, 0.5), "cm"),
        legend.position = 'bottom',
        legend.text=element_text(size=8),
        legend.key.size = unit(0.4, 'cm'),
        legend.title=element_blank())+
  phylum_scale_8+
  ylab('Relative Abundance')+
  ggtitle('Elite Phylum Relative Abundance')

elite_relative_merged_plot
ggsave('images/q30_elite_relative_merged_phylum.pdf',plot=elite_relative_merged_plot,device='pdf',
       useDingbats=FALSE,width=13,height=10,units='cm')
```
## MapQ30 noMQ0 Reads

Reads were also filtered to remove multipmapping reads with a MapQ score of 0, which were still represented in the mapq30 filtered reads.

```{r nomq0_remaining}
euk_remains<-read.table('kraken/euk_barley_props_filtered.txt',sep="\t",header=TRUE) %>%
  filter(Filter != 'nt_mapq30_paired') %>%
  filter(! grepl('minlen',Filter)) %>%
  mutate(Filter = case_when(
    Filter=='nt_mapq10' ~ 'MapQ 10',
    Filter=='nt_mapq15' ~ 'MapQ 15',
    Filter=='nt_mapq20' ~ 'MapQ 20',
    Filter=='nt_mapq25' ~ 'MapQ 25',
    Filter=='nt_mapq30' ~ 'MapQ 30',
    Filter=='mapq_30_nomq0' ~ 'MapQ 30, no MQ 0',
    TRUE ~ (Filter))) %>%
  filter(grepl('MapQ',Filter))

remains.summary <- euk_remains %>%
  filter((grepl('MapQ',Filter))) %>%
  group_by(Sample, Filter) %>%
  summarise(
    euk_sd = sd(Eukaryote),
    euk_props = mean(Eukaryote),
    barley_sd = sd(Barley),
    barley_props = mean(Barley)
  )

barley_plot<-ggplot(remains.summary,aes(Filter,barley_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Barley),
              size=1,position = position_jitter(width = 0.3),colour='darkslategray') + 
  theme_bw()+
  ylim(0,NA)+
  scale_color_startrek()+
  scale_fill_startrek()+
  ylab('Percentage')+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Barley Proportion')

barley_plot

euk_plot<-ggplot(remains.summary,aes(Filter,euk_props)) +
  geom_col(aes(colour=Sample,fill=Sample),alpha=0.7,position = position_dodge(0.8), 
           width = 0.7) +
  geom_point(data=euk_remains, 
              aes(x=Filter,y=Eukaryote),
              size=1,position = position_dodge2(0.8),colour='darkslategray') + 
  theme_bw()+
  ylab('Percentage')+
  scale_color_startrek()+
  scale_fill_startrek()+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1)
        )+
  facet_wrap('Sample')+
  ggtitle('Eukaryote Proportion')
euk_plot
combined_plot<-plot_grid(euk_plot,barley_plot,ncol=1)
combined_plot
ggsave('images/remaining_euk_barley_mapq_nomq0.pdf',plot=combined_plot,device='pdf',width=20,height=20,units='cm',useDingbats=FALSE)
```
```{r phyloseq_prep}
elite_samples_df<-read.csv('elite_map_nomq0.txt',header = TRUE,sep="\t")
bulk_samples_df<-read.csv('bulk_map_nomq0.txt',header = TRUE,sep="\t")

row.names(elite_samples_df)<-elite_samples_df$SampleID
row.names(bulk_samples_df)<-bulk_samples_df$SampleID

bulk_samples_df<-bulk_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))


elite_samples_df<-elite_samples_df %>%
  mutate(Status = case_when(
  Status=='Unfiltered' ~ 'Total',
  Status=='MorexV3 aligned' ~ 'Discarded',
  Status=='MorexV3 unaligned' ~ 'Retained',
  TRUE ~ (Status)))

elite_samples<-row.names(elite_samples_df)
bulk_samples<-row.names(bulk_samples_df)

elite_SAM = sample_data(elite_samples_df, errorIfNULL = T)
bulk_SAM = sample_data(bulk_samples_df, errorIfNULL = T)

ps<-import_biom('kraken/biom/table_mapq30_nomq0.biom.gz')
colnames(tax_table(ps))<-c("Kingdom","Phylum","Class","Order","Family","Genus","Species")
# clean up taxtable entries to remove k__, c__ etc prefixes
tax_table(ps)[, colnames(tax_table(ps))] <- gsub(tax_table(ps)[, colnames(tax_table(ps))], pattern = "[a-z]__", replacement = "")

bact_ps<-subset_taxa(ps,Kingdom=='Bacteria')

elite_bact_ps<-prune_samples(elite_samples,bact_ps)
bulk_bact_ps<-prune_samples(bulk_samples,bact_ps)

elite_bact_ps<-phyloseq(otu_table(elite_bact_ps),tax_table(elite_bact_ps),elite_SAM)
bulk_bact_ps<-phyloseq(otu_table(bulk_bact_ps),tax_table(bulk_bact_ps),bulk_SAM)
```

```{r bray_distance_4}
distmat<-phyloseq::distance(elite_bact_ps,method='bray')
mds<-ordinate(elite_bact_ps,"MDS",distance=distmat)

elite_dist_plot<-plot_ordination(elite_bact_ps,mds,color = 'Status',title = 'Elite Samples Bray-Curtis PCoA')+
  theme_bw()+
  theme(legend.position = 'bottom')+
  scale_color_colorblind()+
  geom_point(size=4)

elite_dist_plot$data$Status=factor(elite_dist_plot$data$Status,
                                   levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired', 'MapQ30,No MQ 0'))
elite_dist_plot<-elite_dist_plot+theme(legend.position='none')

distmat<-phyloseq::distance(bulk_bact_ps,method='bray')
mds<-ordinate(bulk_bact_ps,"MDS",distance=distmat)

bulk_dist_plot<-plot_ordination(bulk_bact_ps,mds,color = 'Status',title = 'Bulk Bray-Curtis PCoA')+
  theme_bw()+
  theme(legend.position = 'bottom')+
  scale_color_colorblind()+
  geom_point(size=4)

bulk_dist_plot$data$Status=factor(bulk_dist_plot$data$Status,
                                   levels=c('Total','Discarded','Retained','MapQ30','MapQ30 Paired', 'MapQ30,No MQ 0'))
dist_plot=bulk_dist_plot+elite_dist_plot & theme(legend.position='bottom',legend.title = element_blank())
dist_plot<-dist_plot+plot_layout(guides = "collect")
ggsave('images/bray4.pdf',plot=dist_plot, device='pdf',width=30,height=15,units='cm',useDingbats=FALSE)
```

```{r adonis_tests}
get_p<-function(cond1,cond2) {
  
  if (cond1 != cond2) {
    df<-meta_distance %>% 
      filter(Status == cond1 | Status == cond2)
    
    dist <- df %>%
      select(all_of(.[['SampleID']])) %>%
      as.dist()
    
    res<-adonis(dist~Status,data=df, permutations=10000)
    return(res[['aov.tab']][['Pr(>F)']][[1]])
} 
  else {
    return(NA)
  }
}

set.seed(1000)

# overall comparison
distmat<-phyloseq::distance(bulk_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(bulk_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res

distmat<-phyloseq::distance(elite_bact_ps,method='bray')
design<-as.data.frame(as.matrix(sample_data(elite_bact_ps)))
res<-adonis(distmat~Status, data=design, permutations = 10000)
res

distance<-as.data.frame(as.matrix(phyloseq::distance(bulk_bact_ps,method='bray')))
distance['SampleID']<-row.names(distance)

meta<-as.data.frame(as.matrix(sample_data(bulk_bact_ps)))
meta_distance<-inner_join(meta,distance,by='SampleID')
all_dist<-meta_distance %>% select(all_of(.[['SampleID']]))

status<-unique(meta_distance$Status)
p_vals<-matrix(nrow=length(status),ncol=length(status))
dimnames(p_vals)=list(
  status,
  status
)

for (cond1 in status) {
  for (cond2 in status) {
    if (cond1 != cond2) {
      p_val<-get_p(cond1,cond2)
      p_vals[cond1,cond2]<-p_val
    }
  }
}

bh<-matrix(p.adjust(p_vals,method='BH'),nrow=length(status),ncol=length(status))

```




```{r final_filterplot}
bulk_fastq_counts<-read.table('bulk_fastq_counts_q30_nomq0.txt',sep="\t",header = TRUE)
bulk_fastq_counts['Group']='Bulk'
bulk_fastq_counts['Replicate']=1:3
elite_fastq_counts<-read.table('elite_fastq_counts_q30_nomq0.txt',sep="\t",header = TRUE)
elite_fastq_counts['Group']='Elite'
elite_fastq_counts['Replicate']=1:3

merged_fastq_counts=rbind(bulk_fastq_counts,elite_fastq_counts) 

merged_fastq_counts<-merged_fastq_counts %>%
  pivot_longer(!c(Group,Replicate),names_to='Filter')


merged_fastq_counts<-merged_fastq_counts %>%
  mutate(Filter = case_when(
  Filter=='unfiltered' ~ 'Total',
  Filter=='mapped' ~ 'Discarded',
  Filter=='unmapped' ~ 'Retained',
  Filter=='mapq_30' ~ 'MapQ30',
  Filter=='mapq_30_nomq0' ~ 'MapQ30,No MapQ 0',
  TRUE ~ (Filter)))

merged_fastq_counts$Filter<-factor(merged_fastq_counts$Filter,levels=c('Total','Discarded','Retained','MapQ30','MapQ30,No MapQ 0'))

count.summary <- merged_fastq_counts %>%
  group_by(Group, Filter) %>%
  filter(! grepl('%',Filter)) %>%
  summarise(
    sd = as.integer(sd(value)),
    count = mean(value)
  )

count_plot<-ggplot(count.summary,aes(Filter,count)) +
  geom_col(aes(colour=Group,fill=Group),alpha=0.7,position = position_dodge(0.8), width = 0.7) +
  geom_point(data=merged_fastq_counts, 
              aes(x=Filter,y=value,colour=Filter),
              size=1,position = position_jitter(width=0.2),colour='darkslategray') + 
  theme_bw()+
  scale_color_startrek()+
  scale_fill_startrek()+
  scale_y_continuous(
    "Mean Read Count",
    labels = scales::label_number_si()
  )+
  theme(legend.position='none',
        axis.title.x = element_blank(),
        axis.text.x = element_text(angle=45, hjust = 1,  vjust = 1))+
  facet_wrap('Group')+
  ggtitle('Filtered Read Counts')
count_plot
ggsave('images/final_filtered_read_counts.pdf',plot=count_plot,device='pdf',useDingbats=FALSE,width=20,height=0,units='cm')
```









