{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "169049dd-0dc6-434b-a2f1-056971fe155b",
   "metadata": {},
   "source": [
    "# Host Filtering Assessment\n",
    "\n",
    "Filtering the barley genome from NT metagenomic reads was observed to bias taxonomic classifications, with a reduction in Firmicutes. Let's try to work out why...\n",
    "\n",
    "## Barley Genome\n",
    "\n",
    "We can use the Morex V3 assembly which includes PacBio HiFi reads, and so is less likely to have bacterial contaminants in the assembly than Morex V2 which was used for the previous analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ba028406-f4f9-49de-a0fb-b77a6db04902",
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib3\n",
    "import shutil\n",
    "import subprocess\n",
    "import os\n",
    "import re\n",
    "from glob import glob\n",
    "import pysam\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "samples={'bulk':['2000','2001','2002'], 'elite':['2023','2024','2025']}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8347da9e-fc11-4c03-b069-0e2205d8e840",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.mkdir('reference')\n",
    "http = urllib3.PoolManager()\n",
    "\n",
    "fasta_url='https://doi.ipk-gatersleben.de/DOI/b2f47dfb-47ff-4114-89ae-bad8dcc515a1/b6e6a2e5-2746-4522-8465-019c8f56df7f/1/DOWNLOAD'\n",
    "annot_url='https://doi.ipk-gatersleben.de/DOI/b2f47dfb-47ff-4114-89ae-bad8dcc515a1/5d16cc17-c37f-417f-855d-c5e72c721f6c/1/DOWNLOAD'\n",
    "fasta_path='reference/Barley_MorexV3_psuedomolecules.fasta'\n",
    "annot_path='reference/Barley_MorexV3_psuedomolecules.gff3'\n",
    "    \n",
    "print('Downloading {}'.format(fasta_url))\n",
    "with http.request('GET',fasta_url, preload_content=False) as resp, open(fasta_path, 'wb') as out_file:\n",
    "    shutil.copyfileobj(resp, out_file)\n",
    "r.release_conn()\n",
    "\n",
    "print('Downloading {}'.format(annot_url))\n",
    "with http.request('GET',annot_url, preload_content=False) as resp, open(annot_path, 'wb') as out_file:\n",
    "    shutil.copyfileobj(resp, out_file)\n",
    "r.release_conn()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be478e92-d7cd-4b7a-86da-87fb747a88ab",
   "metadata": {},
   "source": [
    "Now we can build a BWA index..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e0febb8-e580-4b9c-919f-2cf7d3e9938b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.mkdir('bwa')\n",
    "os.mkdir('logs')\n",
    "bwa_path='bwa/MorexV3'\n",
    "os.symlink('../'+fasta_path,bwa_path)\n",
    "\n",
    "process=subprocess.Popen(['bwa','index',bwa_path],\n",
    "    stdout=subprocess.PIPE,\n",
    "    stderr=subprocess.PIPE)\n",
    "stdout,stderr = process.communicate()\n",
    "\n",
    "with open('logs/bwa_morex_v3_index.out','w') as stdout_file:\n",
    "    stdout_file.write(stdout.decode('UTF-8'))\n",
    "\n",
    "with open('logs/bwa_morex_v3_index.err','w') as stderr_file:\n",
    "    stderr_file.write(stderr.decode('UTF-8'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcdab315-e1ae-42d0-9b76-f834f253a071",
   "metadata": {},
   "source": [
    "...and symlink in the original QC'd fastq files..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac55334e-19c7-4f09-b748-31325dc5150d",
   "metadata": {},
   "outputs": [],
   "source": [
    "files=glob('../GO_enrichment/read_qc/trim_galore/*gz')\n",
    "if not os.path.exists('fastq'):\n",
    "    os.mkdir('fastq')\n",
    "\n",
    "fq_re=re.compile(r'([0-9]{4}_[12])_val_[12].fq.gz')\n",
    "for file in files:\n",
    "    m=re.search(fq_re,file)\n",
    "    os.symlink('../'+file,'fastq/{}.fq.gz'.format(m.group(1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7cd6898-fc37-4d63-824f-0935a9dd609f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat bin/align_and_filter.sh\n",
    "#qsub bin/align_and_filter.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "827770ab-e7f8-4432-8493-83f1ae5000e4",
   "metadata": {},
   "source": [
    "### Mapping Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e313d51b-faa4-4ba2-bde0-98bfad6adb0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2000\n",
      "2001\n",
      "2002\n",
      "2023\n",
      "2024\n",
      "2025\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOStream.flush timed out\n",
      "IOStream.flush timed out\n"
     ]
    }
   ],
   "source": [
    "tot_re=re.compile(r'^([0-9]+) \\+ [0-9]+ in total')\n",
    "map_re=re.compile(r'([0-9]+) \\+ [0-9]+ mapped')\n",
    "\n",
    "labels=[]\n",
    "totals=[]\n",
    "mapped=[]\n",
    "\n",
    "for sample in samples.keys():\n",
    "    for rep in samples[sample]:   \n",
    "        print(rep)\n",
    "        labels.append(rep)\n",
    "        flagstat=pysam.flagstat(\"bwa_alignments/{}.sorted.bam\".format(rep))\n",
    "        m=re.search(tot_re,flagstat)\n",
    "        totals.append(m.group(1))\n",
    "    \n",
    "        m=re.search(map_re,flagstat)\n",
    "        mapped.append(m.group(1))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3304fb-858d-4cc8-ade3-3a68fba62a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "totals=[int(x) for x in totals]\n",
    "mapped=[int(x) for x in mapped]\n",
    "percents=[]\n",
    "for i in range(6):\n",
    "    percents.append((mapped[i]/totals[i])*100)\n",
    "\n",
    "x=np.arange(len(labels))\n",
    "width=0.35\n",
    "\n",
    "plt.rcParams['figure.figsize'] = [12, 6]\n",
    "plt.rcParams['figure.dpi'] = 100\n",
    "\n",
    "fig, (ax1,ax2) = plt.subplots(1,2)\n",
    "rects1 = ax1.bar(x - width/2, totals, width, color=(1, 0.607, 0.314, 1),label='Total Reads')\n",
    "rects2 = ax1.bar(x + width/2, mapped, width, label='Mapped Reads')\n",
    "\n",
    "ax1.set_ylabel('Reads')\n",
    "ax1.set_xticks(x)\n",
    "ax1.set_xticklabels(labels)\n",
    "ax1.ticklabel_format(axis='y',style='sci')\n",
    "\n",
    "rects3 = ax2.bar(x,percents,width*2)\n",
    "ax2.set_xticks(x)\n",
    "ax2.set_xticklabels(labels)\n",
    "ax2.set_ylabel('Mapped Reads (%)')\n",
    "\n",
    "\n",
    "fig.suptitle('Bulk and Elite Read Mapping')\n",
    "fig.legend(loc='upper center',bbox_to_anchor=(0.5,-0.01),ncol=2)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b2c0d04-4ced-4833-830b-4641f374311c",
   "metadata": {},
   "source": [
    "## Kraken\n",
    "\n",
    "We can now classify both the filtered and unfiltered reads with Kraken2 and the kraken2_pluspfp database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "73ae8230-e435-4933-a0a0-2f4805c72a5f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/env bash\n",
      "\n",
      "#$ -pe smp 24\n",
      "#$ -mods l_hard mfree 96G\n",
      "#$ -j y\n",
      "#$ -o logs/$JOB_NAME.$JOB_ID.$TASK_ID\n",
      "#$ -cwd\n",
      "#$ -t 1-12\n",
      "\n",
      "set -e\n",
      "\n",
      "# Runs kraken2 against paired fastq reads, where read filenames start with the sample followed by a '_'.\n",
      "\n",
      "# This script is designed to be run as an array job on an SGE cluster \n",
      "# which has local disk storage available on execution nodesaccessed via $TMPDIR, \n",
      "\n",
      "usage=\"Usage: $0 -i /path/to/input/fastq/directory -o /path/to/output/directory -d path/to/kraken_db\" \n",
      "\n",
      "while getopts \"i:o:d:\" opt; do\n",
      "  case \"${opt}\" in\n",
      "\ti )\n",
      "\t\tin_dir=${OPTARG}\n",
      "\t\t;;\n",
      "\to )\n",
      "\t  out_dir=${OPTARG}\n",
      "\t  ;;\n",
      "\td )\n",
      "\t\tdb=${OPTARG}\n",
      "\t  ;;\n",
      "\t\\? )\n",
      "\t\techo ${usage}\n",
      "\t  ;;\n",
      "\t: )\n",
      "\t\techo \"Invalid option: $OPTARG requires an argument\" 1>&2\n",
      "\t;;\n",
      "  esac\n",
      "done\n",
      "\n",
      "if [[ ! -e \"${in_dir}\" ]]; then\n",
      "\techo \"Specificed input directory (${indir}) does not exist\"\n",
      "\texit\n",
      "fi\n",
      "\n",
      "if [[ ! -e \"${db}\" ]]; then\n",
      "\techo \"Specificed kraken database (${db}) does not exist\"\n",
      "\texit\n",
      "fi\n",
      "\n",
      "if [[ -z \"${SGE_TASK_ID}\" ]]; then\n",
      "  echo \"This script should be executed as an SGE array job\"\n",
      "  exit\n",
      "fi\n",
      "\n",
      "mkdir -p ${out_dir}\n",
      "\n",
      "readarray -t samples < <(ls $in_dir|cut -f1 -d_|uniq)\n",
      "sample=${samples[$SGE_TASK_ID-1]}\n",
      "readarray -t files < <(ls -1 $in_dir/${sample}*|xargs -i basename {})\n",
      "\n",
      "mkdir $TMPDIR/kraken_db\n",
      "cp -v ${db}/* $TMPDIR/kraken_db\n",
      "cp -v ${in_dir}/${sample}* $TMPDIR\n",
      "kraken2 --db $TMPDIR/kraken_db --threads 24 --use-names --report ${out_dir}/${sample}.report.txt \\\n",
      "\t--output ${out_dir}/${sample}.txt --gzip-compressed --paired ${TMPDIR}/${files[0]} ${TMPDIR}/${files[1]}\n",
      "bracken -d ${TMPDIR}/kraken_db -i ${out_dir}/${sample}.report.txt -o ${out_dir}/${sample}.bracken -r 150 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cat bin/kraken.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a238b34-5cfe-46f0-bcdc-7b32f1f9496a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qsub bin/kraken.sh -i fastq -o kraken/nt_unfiltered -d /cluster/db/jabbott/training_data/kraken2_pluspfp_20210127/\n",
    "qsub bin/kraken.sh -i filtered_fastq -o kraken/nt_morex3 -d /cluster/db/jabbott/training_data/kraken2_pluspfp_20210127/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c13d8fd2-3938-44db-b5ae-d7627110dba8",
   "metadata": {},
   "source": [
    "### Create Krona outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ebd5330-7f9f-4c82-beec-a5bb5b04fc2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p kraken/krona\n",
    "cd kraken/nt_unfiltered\n",
    "ktImportTaxonomy -i -m 3 -t 5 -c 2000.report.txt 2001.report.txt  2002.report.txt -o ../krona/nt_unfiltered_bulk.html\n",
    "ktImportTaxonomy -i -m 3 -t 5 -c 2023.report.txt 2024.report.txt  2025.report.txt -o ../krona/nt_unfiltered_elite.html\n",
    "\n",
    "cd ../nt_morex3\n",
    "ktImportTaxonomy -i -m 3 -t 5 -c 2000.report.txt 2001.report.txt  2002.report.txt -o ../krona/nt_morex3_bulk.html\n",
    "ktImportTaxonomy -i -m 3 -t 5 -c 2023.report.txt 2024.report.txt  2025.report.txt -o ../krona/nt_morex3_elite.html\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a26b55-f8b0-4913-a29a-f56defdd8a75",
   "metadata": {},
   "source": [
    "__Results__\n",
    "\n",
    "* [Unfilted Bulk](kraken/krona/nt_unfiltered_bulk.html)\n",
    "* [Unfiltered Elite](kraken/krona/nt_unfiltered_elite.html)\n",
    "* [Morex3 filtered Bulk](kraken/krona/nt_morex3_bulk.html)\n",
    "* [Morex3 filtered Elite](kraken/krona/nt_morex3_elite.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26984a98-a3d1-4a57-a248-b0f6ba57984c",
   "metadata": {},
   "source": [
    "## DESeq2 Analysis\n",
    "\n",
    "DESeq2 was used to determine any statistically significant changes to the community composition introduced by the filtering process, based upon the kraken2 results. See [host_filter.Rmd](host_filter.Rmd) for the code...\n",
    "\n",
    "This only looked at bulk and elite samples, since elite is Morex then this should be optimally filtered by alignment. Bulk as expected mainly shows changes in plant species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1fdf0f5f-b792-4a37-b001-813905124f40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxa</th>\n",
       "      <th>taxid</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>pre.2000</th>\n",
       "      <th>pre.2001</th>\n",
       "      <th>pre.2002</th>\n",
       "      <th>post.2000</th>\n",
       "      <th>post.2001</th>\n",
       "      <th>post.2002</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Medicago</td>\n",
       "      <td>3877</td>\n",
       "      <td>1.183987e+05</td>\n",
       "      <td>-2.397232</td>\n",
       "      <td>0.339698</td>\n",
       "      <td>-7.056959</td>\n",
       "      <td>1.701860e-12</td>\n",
       "      <td>1.825099e-09</td>\n",
       "      <td>73868</td>\n",
       "      <td>52666</td>\n",
       "      <td>92400</td>\n",
       "      <td>105471</td>\n",
       "      <td>74966</td>\n",
       "      <td>128028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Triticum</td>\n",
       "      <td>4564</td>\n",
       "      <td>7.090361e+06</td>\n",
       "      <td>-2.305467</td>\n",
       "      <td>0.331101</td>\n",
       "      <td>-6.963023</td>\n",
       "      <td>3.330473e-12</td>\n",
       "      <td>1.825099e-09</td>\n",
       "      <td>4650548</td>\n",
       "      <td>3379867</td>\n",
       "      <td>5795182</td>\n",
       "      <td>6273404</td>\n",
       "      <td>4520301</td>\n",
       "      <td>7499573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Geosporobacter</td>\n",
       "      <td>390805</td>\n",
       "      <td>1.006955e+02</td>\n",
       "      <td>-1.842569</td>\n",
       "      <td>0.296498</td>\n",
       "      <td>-6.214440</td>\n",
       "      <td>5.150787e-10</td>\n",
       "      <td>1.411316e-07</td>\n",
       "      <td>100</td>\n",
       "      <td>73</td>\n",
       "      <td>85</td>\n",
       "      <td>107</td>\n",
       "      <td>53</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brachypodium</td>\n",
       "      <td>15367</td>\n",
       "      <td>7.370273e+03</td>\n",
       "      <td>-1.809906</td>\n",
       "      <td>0.314330</td>\n",
       "      <td>-5.757985</td>\n",
       "      <td>8.512398e-09</td>\n",
       "      <td>1.865918e-06</td>\n",
       "      <td>6554</td>\n",
       "      <td>4821</td>\n",
       "      <td>7625</td>\n",
       "      <td>6308</td>\n",
       "      <td>4305</td>\n",
       "      <td>7216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mogibacterium</td>\n",
       "      <td>86331</td>\n",
       "      <td>1.784931e+02</td>\n",
       "      <td>-1.458128</td>\n",
       "      <td>0.325623</td>\n",
       "      <td>-4.477966</td>\n",
       "      <td>7.535773e-06</td>\n",
       "      <td>6.353237e-04</td>\n",
       "      <td>197</td>\n",
       "      <td>143</td>\n",
       "      <td>215</td>\n",
       "      <td>180</td>\n",
       "      <td>85</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Setaria</td>\n",
       "      <td>4554</td>\n",
       "      <td>4.825621e+03</td>\n",
       "      <td>-1.424191</td>\n",
       "      <td>0.288600</td>\n",
       "      <td>-4.934821</td>\n",
       "      <td>8.022439e-07</td>\n",
       "      <td>1.099074e-04</td>\n",
       "      <td>5364</td>\n",
       "      <td>4045</td>\n",
       "      <td>5840</td>\n",
       "      <td>3829</td>\n",
       "      <td>2749</td>\n",
       "      <td>4353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zea</td>\n",
       "      <td>4575</td>\n",
       "      <td>8.038816e+03</td>\n",
       "      <td>-1.372550</td>\n",
       "      <td>0.291772</td>\n",
       "      <td>-4.704185</td>\n",
       "      <td>2.548825e-06</td>\n",
       "      <td>3.103903e-04</td>\n",
       "      <td>9249</td>\n",
       "      <td>6876</td>\n",
       "      <td>9962</td>\n",
       "      <td>6485</td>\n",
       "      <td>4396</td>\n",
       "      <td>7177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Panicum</td>\n",
       "      <td>4539</td>\n",
       "      <td>4.302652e+03</td>\n",
       "      <td>-1.356234</td>\n",
       "      <td>0.271575</td>\n",
       "      <td>-4.993949</td>\n",
       "      <td>5.915698e-07</td>\n",
       "      <td>9.262293e-05</td>\n",
       "      <td>5028</td>\n",
       "      <td>3773</td>\n",
       "      <td>5286</td>\n",
       "      <td>3561</td>\n",
       "      <td>2409</td>\n",
       "      <td>3694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Homo</td>\n",
       "      <td>9605</td>\n",
       "      <td>1.823806e+05</td>\n",
       "      <td>-1.244964</td>\n",
       "      <td>0.199390</td>\n",
       "      <td>-6.243862</td>\n",
       "      <td>4.268968e-10</td>\n",
       "      <td>1.411316e-07</td>\n",
       "      <td>209346</td>\n",
       "      <td>210294</td>\n",
       "      <td>208995</td>\n",
       "      <td>132190</td>\n",
       "      <td>131857</td>\n",
       "      <td>135838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Sorghum</td>\n",
       "      <td>4557</td>\n",
       "      <td>4.014148e+03</td>\n",
       "      <td>-1.236504</td>\n",
       "      <td>0.273783</td>\n",
       "      <td>-4.516362</td>\n",
       "      <td>6.291111e-06</td>\n",
       "      <td>6.152143e-04</td>\n",
       "      <td>4918</td>\n",
       "      <td>3834</td>\n",
       "      <td>5179</td>\n",
       "      <td>3082</td>\n",
       "      <td>2233</td>\n",
       "      <td>3453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Alphabaculovirus</td>\n",
       "      <td>558016</td>\n",
       "      <td>8.174654e+01</td>\n",
       "      <td>-1.230546</td>\n",
       "      <td>0.269452</td>\n",
       "      <td>-4.566855</td>\n",
       "      <td>4.950968e-06</td>\n",
       "      <td>5.426261e-04</td>\n",
       "      <td>91</td>\n",
       "      <td>87</td>\n",
       "      <td>105</td>\n",
       "      <td>68</td>\n",
       "      <td>51</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Sugiyamaella</td>\n",
       "      <td>410829</td>\n",
       "      <td>1.288534e+02</td>\n",
       "      <td>-1.088068</td>\n",
       "      <td>0.241693</td>\n",
       "      <td>-4.501867</td>\n",
       "      <td>6.735923e-06</td>\n",
       "      <td>6.152143e-04</td>\n",
       "      <td>188</td>\n",
       "      <td>131</td>\n",
       "      <td>165</td>\n",
       "      <td>109</td>\n",
       "      <td>72</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Oryza</td>\n",
       "      <td>4527</td>\n",
       "      <td>5.097811e+03</td>\n",
       "      <td>-1.030991</td>\n",
       "      <td>0.252589</td>\n",
       "      <td>-4.081702</td>\n",
       "      <td>4.470710e-05</td>\n",
       "      <td>2.449949e-03</td>\n",
       "      <td>7023</td>\n",
       "      <td>5634</td>\n",
       "      <td>6892</td>\n",
       "      <td>3814</td>\n",
       "      <td>2726</td>\n",
       "      <td>4127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bathycoccus</td>\n",
       "      <td>41874</td>\n",
       "      <td>1.548521e+02</td>\n",
       "      <td>-1.022473</td>\n",
       "      <td>0.259170</td>\n",
       "      <td>-3.945179</td>\n",
       "      <td>7.974051e-05</td>\n",
       "      <td>3.972527e-03</td>\n",
       "      <td>219</td>\n",
       "      <td>170</td>\n",
       "      <td>208</td>\n",
       "      <td>128</td>\n",
       "      <td>79</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Nicotiana</td>\n",
       "      <td>4085</td>\n",
       "      <td>3.603619e+03</td>\n",
       "      <td>-0.974647</td>\n",
       "      <td>0.229918</td>\n",
       "      <td>-4.239100</td>\n",
       "      <td>2.244180e-05</td>\n",
       "      <td>1.382136e-03</td>\n",
       "      <td>4982</td>\n",
       "      <td>4420</td>\n",
       "      <td>4767</td>\n",
       "      <td>2586</td>\n",
       "      <td>2029</td>\n",
       "      <td>2822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Cannabis</td>\n",
       "      <td>3482</td>\n",
       "      <td>1.535299e+03</td>\n",
       "      <td>-0.880005</td>\n",
       "      <td>0.210731</td>\n",
       "      <td>-4.175958</td>\n",
       "      <td>2.967350e-05</td>\n",
       "      <td>1.711692e-03</td>\n",
       "      <td>2209</td>\n",
       "      <td>1976</td>\n",
       "      <td>2115</td>\n",
       "      <td>1070</td>\n",
       "      <td>899</td>\n",
       "      <td>1132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Camelina</td>\n",
       "      <td>71323</td>\n",
       "      <td>1.815030e+03</td>\n",
       "      <td>-0.871766</td>\n",
       "      <td>0.224274</td>\n",
       "      <td>-3.887068</td>\n",
       "      <td>1.014625e-04</td>\n",
       "      <td>4.633455e-03</td>\n",
       "      <td>2657</td>\n",
       "      <td>2224</td>\n",
       "      <td>2599</td>\n",
       "      <td>1330</td>\n",
       "      <td>991</td>\n",
       "      <td>1347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Methylobacillus</td>\n",
       "      <td>404</td>\n",
       "      <td>2.262923e+02</td>\n",
       "      <td>-0.863600</td>\n",
       "      <td>0.203846</td>\n",
       "      <td>-4.236537</td>\n",
       "      <td>2.269932e-05</td>\n",
       "      <td>1.382136e-03</td>\n",
       "      <td>400</td>\n",
       "      <td>273</td>\n",
       "      <td>277</td>\n",
       "      <td>179</td>\n",
       "      <td>126</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Cynara</td>\n",
       "      <td>4264</td>\n",
       "      <td>1.662365e+03</td>\n",
       "      <td>-0.824695</td>\n",
       "      <td>0.189107</td>\n",
       "      <td>-4.360994</td>\n",
       "      <td>1.294732e-05</td>\n",
       "      <td>1.013590e-03</td>\n",
       "      <td>2494</td>\n",
       "      <td>2253</td>\n",
       "      <td>2256</td>\n",
       "      <td>1136</td>\n",
       "      <td>1010</td>\n",
       "      <td>1169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Juglans</td>\n",
       "      <td>16718</td>\n",
       "      <td>1.650576e+03</td>\n",
       "      <td>-0.809260</td>\n",
       "      <td>0.200245</td>\n",
       "      <td>-4.041356</td>\n",
       "      <td>5.314310e-05</td>\n",
       "      <td>2.773564e-03</td>\n",
       "      <td>2419</td>\n",
       "      <td>2230</td>\n",
       "      <td>2335</td>\n",
       "      <td>1124</td>\n",
       "      <td>990</td>\n",
       "      <td>1164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Gossypium</td>\n",
       "      <td>3633</td>\n",
       "      <td>6.633990e+03</td>\n",
       "      <td>-0.804386</td>\n",
       "      <td>0.206123</td>\n",
       "      <td>-3.902458</td>\n",
       "      <td>9.522073e-05</td>\n",
       "      <td>4.537475e-03</td>\n",
       "      <td>9895</td>\n",
       "      <td>8965</td>\n",
       "      <td>9307</td>\n",
       "      <td>4772</td>\n",
       "      <td>3657</td>\n",
       "      <td>4761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Helianthus</td>\n",
       "      <td>4231</td>\n",
       "      <td>3.544506e+03</td>\n",
       "      <td>-0.788023</td>\n",
       "      <td>0.185960</td>\n",
       "      <td>-4.237584</td>\n",
       "      <td>2.259376e-05</td>\n",
       "      <td>1.382136e-03</td>\n",
       "      <td>5267</td>\n",
       "      <td>5042</td>\n",
       "      <td>4843</td>\n",
       "      <td>2538</td>\n",
       "      <td>2057</td>\n",
       "      <td>2433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Theobroma</td>\n",
       "      <td>3640</td>\n",
       "      <td>8.116086e+02</td>\n",
       "      <td>-0.762328</td>\n",
       "      <td>0.175529</td>\n",
       "      <td>-4.343030</td>\n",
       "      <td>1.405308e-05</td>\n",
       "      <td>1.026812e-03</td>\n",
       "      <td>1326</td>\n",
       "      <td>1108</td>\n",
       "      <td>1095</td>\n",
       "      <td>565</td>\n",
       "      <td>477</td>\n",
       "      <td>555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Prunus</td>\n",
       "      <td>3754</td>\n",
       "      <td>1.283711e+03</td>\n",
       "      <td>-0.758156</td>\n",
       "      <td>0.198753</td>\n",
       "      <td>-3.814558</td>\n",
       "      <td>1.364268e-04</td>\n",
       "      <td>5.980950e-03</td>\n",
       "      <td>1986</td>\n",
       "      <td>1720</td>\n",
       "      <td>1861</td>\n",
       "      <td>940</td>\n",
       "      <td>700</td>\n",
       "      <td>889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Brassica</td>\n",
       "      <td>3705</td>\n",
       "      <td>3.048838e+03</td>\n",
       "      <td>-0.757898</td>\n",
       "      <td>0.241775</td>\n",
       "      <td>-3.134727</td>\n",
       "      <td>1.720142e-03</td>\n",
       "      <td>4.284716e-02</td>\n",
       "      <td>4400</td>\n",
       "      <td>4099</td>\n",
       "      <td>4655</td>\n",
       "      <td>2065</td>\n",
       "      <td>1642</td>\n",
       "      <td>2258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Sesamum</td>\n",
       "      <td>4181</td>\n",
       "      <td>9.738222e+02</td>\n",
       "      <td>-0.751392</td>\n",
       "      <td>0.206009</td>\n",
       "      <td>-3.647375</td>\n",
       "      <td>2.649329e-04</td>\n",
       "      <td>9.954971e-03</td>\n",
       "      <td>1477</td>\n",
       "      <td>1318</td>\n",
       "      <td>1433</td>\n",
       "      <td>684</td>\n",
       "      <td>544</td>\n",
       "      <td>682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Solanum</td>\n",
       "      <td>4107</td>\n",
       "      <td>4.090971e+03</td>\n",
       "      <td>-0.725773</td>\n",
       "      <td>0.213394</td>\n",
       "      <td>-3.401088</td>\n",
       "      <td>6.711814e-04</td>\n",
       "      <td>1.988148e-02</td>\n",
       "      <td>6365</td>\n",
       "      <td>5670</td>\n",
       "      <td>5950</td>\n",
       "      <td>2903</td>\n",
       "      <td>2156</td>\n",
       "      <td>2912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Arachis</td>\n",
       "      <td>3817</td>\n",
       "      <td>4.136350e+03</td>\n",
       "      <td>-0.721334</td>\n",
       "      <td>0.201751</td>\n",
       "      <td>-3.575362</td>\n",
       "      <td>3.497432e-04</td>\n",
       "      <td>1.208918e-02</td>\n",
       "      <td>6742</td>\n",
       "      <td>5796</td>\n",
       "      <td>5748</td>\n",
       "      <td>2933</td>\n",
       "      <td>2195</td>\n",
       "      <td>2924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Manihot</td>\n",
       "      <td>3982</td>\n",
       "      <td>1.125732e+03</td>\n",
       "      <td>-0.709539</td>\n",
       "      <td>0.141934</td>\n",
       "      <td>-4.999089</td>\n",
       "      <td>5.760173e-07</td>\n",
       "      <td>9.262293e-05</td>\n",
       "      <td>1833</td>\n",
       "      <td>1636</td>\n",
       "      <td>1532</td>\n",
       "      <td>824</td>\n",
       "      <td>682</td>\n",
       "      <td>693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Capsicum</td>\n",
       "      <td>4071</td>\n",
       "      <td>5.957803e+03</td>\n",
       "      <td>-0.700180</td>\n",
       "      <td>0.205189</td>\n",
       "      <td>-3.412357</td>\n",
       "      <td>6.440365e-04</td>\n",
       "      <td>1.960733e-02</td>\n",
       "      <td>9366</td>\n",
       "      <td>8470</td>\n",
       "      <td>8649</td>\n",
       "      <td>4096</td>\n",
       "      <td>3236</td>\n",
       "      <td>4185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Glycine</td>\n",
       "      <td>3846</td>\n",
       "      <td>2.529422e+03</td>\n",
       "      <td>-0.685846</td>\n",
       "      <td>0.188412</td>\n",
       "      <td>-3.640140</td>\n",
       "      <td>2.724901e-04</td>\n",
       "      <td>9.954971e-03</td>\n",
       "      <td>3902</td>\n",
       "      <td>3739</td>\n",
       "      <td>3654</td>\n",
       "      <td>1708</td>\n",
       "      <td>1474</td>\n",
       "      <td>1693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Marinobacter</td>\n",
       "      <td>2742</td>\n",
       "      <td>2.381718e+03</td>\n",
       "      <td>-0.645882</td>\n",
       "      <td>0.172190</td>\n",
       "      <td>-3.750985</td>\n",
       "      <td>1.761411e-04</td>\n",
       "      <td>7.425023e-03</td>\n",
       "      <td>4264</td>\n",
       "      <td>3259</td>\n",
       "      <td>3401</td>\n",
       "      <td>1726</td>\n",
       "      <td>1285</td>\n",
       "      <td>1554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Punica</td>\n",
       "      <td>22662</td>\n",
       "      <td>7.755750e+02</td>\n",
       "      <td>-0.635361</td>\n",
       "      <td>0.200694</td>\n",
       "      <td>-3.165811</td>\n",
       "      <td>1.546512e-03</td>\n",
       "      <td>4.107057e-02</td>\n",
       "      <td>1215</td>\n",
       "      <td>1160</td>\n",
       "      <td>1161</td>\n",
       "      <td>541</td>\n",
       "      <td>419</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>3701</td>\n",
       "      <td>3.314352e+02</td>\n",
       "      <td>-0.621402</td>\n",
       "      <td>0.173918</td>\n",
       "      <td>-3.572960</td>\n",
       "      <td>3.529687e-04</td>\n",
       "      <td>1.208918e-02</td>\n",
       "      <td>549</td>\n",
       "      <td>511</td>\n",
       "      <td>466</td>\n",
       "      <td>232</td>\n",
       "      <td>193</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Ananas</td>\n",
       "      <td>4614</td>\n",
       "      <td>1.211435e+03</td>\n",
       "      <td>-0.617552</td>\n",
       "      <td>0.174573</td>\n",
       "      <td>-3.537493</td>\n",
       "      <td>4.039451e-04</td>\n",
       "      <td>1.341587e-02</td>\n",
       "      <td>2019</td>\n",
       "      <td>1784</td>\n",
       "      <td>1784</td>\n",
       "      <td>833</td>\n",
       "      <td>678</td>\n",
       "      <td>787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Olea</td>\n",
       "      <td>4145</td>\n",
       "      <td>2.032805e+03</td>\n",
       "      <td>-0.612016</td>\n",
       "      <td>0.200166</td>\n",
       "      <td>-3.057537</td>\n",
       "      <td>2.231640e-03</td>\n",
       "      <td>4.991586e-02</td>\n",
       "      <td>3248</td>\n",
       "      <td>3314</td>\n",
       "      <td>2813</td>\n",
       "      <td>1360</td>\n",
       "      <td>1128</td>\n",
       "      <td>1353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Lupinus</td>\n",
       "      <td>3869</td>\n",
       "      <td>1.556455e+03</td>\n",
       "      <td>-0.570761</td>\n",
       "      <td>0.180580</td>\n",
       "      <td>-3.160706</td>\n",
       "      <td>1.573872e-03</td>\n",
       "      <td>4.107057e-02</td>\n",
       "      <td>2562</td>\n",
       "      <td>2477</td>\n",
       "      <td>2267</td>\n",
       "      <td>1033</td>\n",
       "      <td>882</td>\n",
       "      <td>996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Malus</td>\n",
       "      <td>3749</td>\n",
       "      <td>1.038195e+03</td>\n",
       "      <td>-0.562612</td>\n",
       "      <td>0.181271</td>\n",
       "      <td>-3.103699</td>\n",
       "      <td>1.911175e-03</td>\n",
       "      <td>4.553582e-02</td>\n",
       "      <td>1823</td>\n",
       "      <td>1579</td>\n",
       "      <td>1509</td>\n",
       "      <td>729</td>\n",
       "      <td>542</td>\n",
       "      <td>672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Aerosticca</td>\n",
       "      <td>2707020</td>\n",
       "      <td>2.254461e+03</td>\n",
       "      <td>-0.550183</td>\n",
       "      <td>0.149028</td>\n",
       "      <td>-3.691816</td>\n",
       "      <td>2.226585e-04</td>\n",
       "      <td>9.018467e-03</td>\n",
       "      <td>4096</td>\n",
       "      <td>3104</td>\n",
       "      <td>3536</td>\n",
       "      <td>1630</td>\n",
       "      <td>1281</td>\n",
       "      <td>1315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Elaeis</td>\n",
       "      <td>51952</td>\n",
       "      <td>2.199089e+03</td>\n",
       "      <td>-0.536760</td>\n",
       "      <td>0.175316</td>\n",
       "      <td>-3.061672</td>\n",
       "      <td>2.201046e-03</td>\n",
       "      <td>4.991586e-02</td>\n",
       "      <td>3840</td>\n",
       "      <td>3466</td>\n",
       "      <td>3197</td>\n",
       "      <td>1425</td>\n",
       "      <td>1234</td>\n",
       "      <td>1410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Cicer</td>\n",
       "      <td>3826</td>\n",
       "      <td>1.152033e+03</td>\n",
       "      <td>-0.534877</td>\n",
       "      <td>0.168789</td>\n",
       "      <td>-3.168907</td>\n",
       "      <td>1.530133e-03</td>\n",
       "      <td>4.107057e-02</td>\n",
       "      <td>1887</td>\n",
       "      <td>1931</td>\n",
       "      <td>1665</td>\n",
       "      <td>806</td>\n",
       "      <td>666</td>\n",
       "      <td>674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Vitis</td>\n",
       "      <td>3603</td>\n",
       "      <td>1.454077e+03</td>\n",
       "      <td>-0.488835</td>\n",
       "      <td>0.158845</td>\n",
       "      <td>-3.077424</td>\n",
       "      <td>2.087983e-03</td>\n",
       "      <td>4.868998e-02</td>\n",
       "      <td>2585</td>\n",
       "      <td>2364</td>\n",
       "      <td>2133</td>\n",
       "      <td>999</td>\n",
       "      <td>793</td>\n",
       "      <td>877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Frateuria</td>\n",
       "      <td>70411</td>\n",
       "      <td>6.357506e+02</td>\n",
       "      <td>-0.463485</td>\n",
       "      <td>0.147395</td>\n",
       "      <td>-3.144499</td>\n",
       "      <td>1.663717e-03</td>\n",
       "      <td>4.240545e-02</td>\n",
       "      <td>1205</td>\n",
       "      <td>923</td>\n",
       "      <td>1011</td>\n",
       "      <td>457</td>\n",
       "      <td>359</td>\n",
       "      <td>349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Rubrivivax</td>\n",
       "      <td>28067</td>\n",
       "      <td>2.103171e+03</td>\n",
       "      <td>-0.444890</td>\n",
       "      <td>0.120792</td>\n",
       "      <td>-3.683118</td>\n",
       "      <td>2.303988e-04</td>\n",
       "      <td>9.018467e-03</td>\n",
       "      <td>4035</td>\n",
       "      <td>3372</td>\n",
       "      <td>3068</td>\n",
       "      <td>1494</td>\n",
       "      <td>1214</td>\n",
       "      <td>1126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Dyella</td>\n",
       "      <td>231454</td>\n",
       "      <td>1.127381e+04</td>\n",
       "      <td>-0.441464</td>\n",
       "      <td>0.128542</td>\n",
       "      <td>-3.434396</td>\n",
       "      <td>5.938769e-04</td>\n",
       "      <td>1.859683e-02</td>\n",
       "      <td>20572</td>\n",
       "      <td>17422</td>\n",
       "      <td>17968</td>\n",
       "      <td>7828</td>\n",
       "      <td>6385</td>\n",
       "      <td>6268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Archangium</td>\n",
       "      <td>47</td>\n",
       "      <td>2.688560e+03</td>\n",
       "      <td>-0.396997</td>\n",
       "      <td>0.127254</td>\n",
       "      <td>-3.119731</td>\n",
       "      <td>1.810160e-03</td>\n",
       "      <td>4.408745e-02</td>\n",
       "      <td>4997</td>\n",
       "      <td>4345</td>\n",
       "      <td>4246</td>\n",
       "      <td>1806</td>\n",
       "      <td>1617</td>\n",
       "      <td>1399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Luteibacter</td>\n",
       "      <td>242605</td>\n",
       "      <td>3.244481e+03</td>\n",
       "      <td>-0.394462</td>\n",
       "      <td>0.114371</td>\n",
       "      <td>-3.448960</td>\n",
       "      <td>5.627502e-04</td>\n",
       "      <td>1.814042e-02</td>\n",
       "      <td>6267</td>\n",
       "      <td>5197</td>\n",
       "      <td>5000</td>\n",
       "      <td>2261</td>\n",
       "      <td>1884</td>\n",
       "      <td>1683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Alteromonas</td>\n",
       "      <td>226</td>\n",
       "      <td>3.654832e+02</td>\n",
       "      <td>0.608987</td>\n",
       "      <td>0.190844</td>\n",
       "      <td>3.191017</td>\n",
       "      <td>1.417727e-03</td>\n",
       "      <td>3.984177e-02</td>\n",
       "      <td>883</td>\n",
       "      <td>900</td>\n",
       "      <td>791</td>\n",
       "      <td>190</td>\n",
       "      <td>146</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Paraglaciecola</td>\n",
       "      <td>1621534</td>\n",
       "      <td>1.093325e+02</td>\n",
       "      <td>0.742545</td>\n",
       "      <td>0.222151</td>\n",
       "      <td>3.342523</td>\n",
       "      <td>8.302049e-04</td>\n",
       "      <td>2.394486e-02</td>\n",
       "      <td>293</td>\n",
       "      <td>260</td>\n",
       "      <td>248</td>\n",
       "      <td>47</td>\n",
       "      <td>46</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                taxa    taxid      baseMean  log2FoldChange     lfcSE  \\\n",
       "0           Medicago     3877  1.183987e+05       -2.397232  0.339698   \n",
       "1           Triticum     4564  7.090361e+06       -2.305467  0.331101   \n",
       "2     Geosporobacter   390805  1.006955e+02       -1.842569  0.296498   \n",
       "3       Brachypodium    15367  7.370273e+03       -1.809906  0.314330   \n",
       "4      Mogibacterium    86331  1.784931e+02       -1.458128  0.325623   \n",
       "5            Setaria     4554  4.825621e+03       -1.424191  0.288600   \n",
       "6                Zea     4575  8.038816e+03       -1.372550  0.291772   \n",
       "7            Panicum     4539  4.302652e+03       -1.356234  0.271575   \n",
       "8               Homo     9605  1.823806e+05       -1.244964  0.199390   \n",
       "9            Sorghum     4557  4.014148e+03       -1.236504  0.273783   \n",
       "10  Alphabaculovirus   558016  8.174654e+01       -1.230546  0.269452   \n",
       "11      Sugiyamaella   410829  1.288534e+02       -1.088068  0.241693   \n",
       "12             Oryza     4527  5.097811e+03       -1.030991  0.252589   \n",
       "13       Bathycoccus    41874  1.548521e+02       -1.022473  0.259170   \n",
       "14         Nicotiana     4085  3.603619e+03       -0.974647  0.229918   \n",
       "15          Cannabis     3482  1.535299e+03       -0.880005  0.210731   \n",
       "16          Camelina    71323  1.815030e+03       -0.871766  0.224274   \n",
       "17   Methylobacillus      404  2.262923e+02       -0.863600  0.203846   \n",
       "18            Cynara     4264  1.662365e+03       -0.824695  0.189107   \n",
       "19           Juglans    16718  1.650576e+03       -0.809260  0.200245   \n",
       "20         Gossypium     3633  6.633990e+03       -0.804386  0.206123   \n",
       "21        Helianthus     4231  3.544506e+03       -0.788023  0.185960   \n",
       "22         Theobroma     3640  8.116086e+02       -0.762328  0.175529   \n",
       "23            Prunus     3754  1.283711e+03       -0.758156  0.198753   \n",
       "24          Brassica     3705  3.048838e+03       -0.757898  0.241775   \n",
       "25           Sesamum     4181  9.738222e+02       -0.751392  0.206009   \n",
       "26           Solanum     4107  4.090971e+03       -0.725773  0.213394   \n",
       "27           Arachis     3817  4.136350e+03       -0.721334  0.201751   \n",
       "28           Manihot     3982  1.125732e+03       -0.709539  0.141934   \n",
       "29          Capsicum     4071  5.957803e+03       -0.700180  0.205189   \n",
       "30           Glycine     3846  2.529422e+03       -0.685846  0.188412   \n",
       "31      Marinobacter     2742  2.381718e+03       -0.645882  0.172190   \n",
       "32            Punica    22662  7.755750e+02       -0.635361  0.200694   \n",
       "33       Arabidopsis     3701  3.314352e+02       -0.621402  0.173918   \n",
       "34            Ananas     4614  1.211435e+03       -0.617552  0.174573   \n",
       "35              Olea     4145  2.032805e+03       -0.612016  0.200166   \n",
       "36           Lupinus     3869  1.556455e+03       -0.570761  0.180580   \n",
       "37             Malus     3749  1.038195e+03       -0.562612  0.181271   \n",
       "38        Aerosticca  2707020  2.254461e+03       -0.550183  0.149028   \n",
       "39            Elaeis    51952  2.199089e+03       -0.536760  0.175316   \n",
       "40             Cicer     3826  1.152033e+03       -0.534877  0.168789   \n",
       "41             Vitis     3603  1.454077e+03       -0.488835  0.158845   \n",
       "42         Frateuria    70411  6.357506e+02       -0.463485  0.147395   \n",
       "43        Rubrivivax    28067  2.103171e+03       -0.444890  0.120792   \n",
       "44            Dyella   231454  1.127381e+04       -0.441464  0.128542   \n",
       "45        Archangium       47  2.688560e+03       -0.396997  0.127254   \n",
       "46       Luteibacter   242605  3.244481e+03       -0.394462  0.114371   \n",
       "47       Alteromonas      226  3.654832e+02        0.608987  0.190844   \n",
       "48    Paraglaciecola  1621534  1.093325e+02        0.742545  0.222151   \n",
       "\n",
       "        stat        pvalue          padj  pre.2000  pre.2001  pre.2002  \\\n",
       "0  -7.056959  1.701860e-12  1.825099e-09     73868     52666     92400   \n",
       "1  -6.963023  3.330473e-12  1.825099e-09   4650548   3379867   5795182   \n",
       "2  -6.214440  5.150787e-10  1.411316e-07       100        73        85   \n",
       "3  -5.757985  8.512398e-09  1.865918e-06      6554      4821      7625   \n",
       "4  -4.477966  7.535773e-06  6.353237e-04       197       143       215   \n",
       "5  -4.934821  8.022439e-07  1.099074e-04      5364      4045      5840   \n",
       "6  -4.704185  2.548825e-06  3.103903e-04      9249      6876      9962   \n",
       "7  -4.993949  5.915698e-07  9.262293e-05      5028      3773      5286   \n",
       "8  -6.243862  4.268968e-10  1.411316e-07    209346    210294    208995   \n",
       "9  -4.516362  6.291111e-06  6.152143e-04      4918      3834      5179   \n",
       "10 -4.566855  4.950968e-06  5.426261e-04        91        87       105   \n",
       "11 -4.501867  6.735923e-06  6.152143e-04       188       131       165   \n",
       "12 -4.081702  4.470710e-05  2.449949e-03      7023      5634      6892   \n",
       "13 -3.945179  7.974051e-05  3.972527e-03       219       170       208   \n",
       "14 -4.239100  2.244180e-05  1.382136e-03      4982      4420      4767   \n",
       "15 -4.175958  2.967350e-05  1.711692e-03      2209      1976      2115   \n",
       "16 -3.887068  1.014625e-04  4.633455e-03      2657      2224      2599   \n",
       "17 -4.236537  2.269932e-05  1.382136e-03       400       273       277   \n",
       "18 -4.360994  1.294732e-05  1.013590e-03      2494      2253      2256   \n",
       "19 -4.041356  5.314310e-05  2.773564e-03      2419      2230      2335   \n",
       "20 -3.902458  9.522073e-05  4.537475e-03      9895      8965      9307   \n",
       "21 -4.237584  2.259376e-05  1.382136e-03      5267      5042      4843   \n",
       "22 -4.343030  1.405308e-05  1.026812e-03      1326      1108      1095   \n",
       "23 -3.814558  1.364268e-04  5.980950e-03      1986      1720      1861   \n",
       "24 -3.134727  1.720142e-03  4.284716e-02      4400      4099      4655   \n",
       "25 -3.647375  2.649329e-04  9.954971e-03      1477      1318      1433   \n",
       "26 -3.401088  6.711814e-04  1.988148e-02      6365      5670      5950   \n",
       "27 -3.575362  3.497432e-04  1.208918e-02      6742      5796      5748   \n",
       "28 -4.999089  5.760173e-07  9.262293e-05      1833      1636      1532   \n",
       "29 -3.412357  6.440365e-04  1.960733e-02      9366      8470      8649   \n",
       "30 -3.640140  2.724901e-04  9.954971e-03      3902      3739      3654   \n",
       "31 -3.750985  1.761411e-04  7.425023e-03      4264      3259      3401   \n",
       "32 -3.165811  1.546512e-03  4.107057e-02      1215      1160      1161   \n",
       "33 -3.572960  3.529687e-04  1.208918e-02       549       511       466   \n",
       "34 -3.537493  4.039451e-04  1.341587e-02      2019      1784      1784   \n",
       "35 -3.057537  2.231640e-03  4.991586e-02      3248      3314      2813   \n",
       "36 -3.160706  1.573872e-03  4.107057e-02      2562      2477      2267   \n",
       "37 -3.103699  1.911175e-03  4.553582e-02      1823      1579      1509   \n",
       "38 -3.691816  2.226585e-04  9.018467e-03      4096      3104      3536   \n",
       "39 -3.061672  2.201046e-03  4.991586e-02      3840      3466      3197   \n",
       "40 -3.168907  1.530133e-03  4.107057e-02      1887      1931      1665   \n",
       "41 -3.077424  2.087983e-03  4.868998e-02      2585      2364      2133   \n",
       "42 -3.144499  1.663717e-03  4.240545e-02      1205       923      1011   \n",
       "43 -3.683118  2.303988e-04  9.018467e-03      4035      3372      3068   \n",
       "44 -3.434396  5.938769e-04  1.859683e-02     20572     17422     17968   \n",
       "45 -3.119731  1.810160e-03  4.408745e-02      4997      4345      4246   \n",
       "46 -3.448960  5.627502e-04  1.814042e-02      6267      5197      5000   \n",
       "47  3.191017  1.417727e-03  3.984177e-02       883       900       791   \n",
       "48  3.342523  8.302049e-04  2.394486e-02       293       260       248   \n",
       "\n",
       "    post.2000  post.2001  post.2002  \n",
       "0      105471      74966     128028  \n",
       "1     6273404    4520301    7499573  \n",
       "2         107         53         89  \n",
       "3        6308       4305       7216  \n",
       "4         180         85        149  \n",
       "5        3829       2749       4353  \n",
       "6        6485       4396       7177  \n",
       "7        3561       2409       3694  \n",
       "8      132190     131857     135838  \n",
       "9        3082       2233       3453  \n",
       "10         68         51         61  \n",
       "11        109         72         95  \n",
       "12       3814       2726       4127  \n",
       "13        128         79        119  \n",
       "14       2586       2029       2822  \n",
       "15       1070        899       1132  \n",
       "16       1330        991       1347  \n",
       "17        179        126        155  \n",
       "18       1136       1010       1169  \n",
       "19       1124        990       1164  \n",
       "20       4772       3657       4761  \n",
       "21       2538       2057       2433  \n",
       "22        565        477        555  \n",
       "23        940        700        889  \n",
       "24       2065       1642       2258  \n",
       "25        684        544        682  \n",
       "26       2903       2156       2912  \n",
       "27       2933       2195       2924  \n",
       "28        824        682        693  \n",
       "29       4096       3236       4185  \n",
       "30       1708       1474       1693  \n",
       "31       1726       1285       1554  \n",
       "32        541        419        518  \n",
       "33        232        193        206  \n",
       "34        833        678        787  \n",
       "35       1360       1128       1353  \n",
       "36       1033        882        996  \n",
       "37        729        542        672  \n",
       "38       1630       1281       1315  \n",
       "39       1425       1234       1410  \n",
       "40        806        666        674  \n",
       "41        999        793        877  \n",
       "42        457        359        349  \n",
       "43       1494       1214       1126  \n",
       "44       7828       6385       6268  \n",
       "45       1806       1617       1399  \n",
       "46       2261       1884       1683  \n",
       "47        190        146        124  \n",
       "48         47         46         36  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bulk_deseq=pd.read_csv('deseq2/bulk.txt',sep=\"\\t\")\n",
    "bulk_deseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1afa5d45-668e-44fa-92fa-38de440e2907",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxa</th>\n",
       "      <th>taxid</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>pre.2023</th>\n",
       "      <th>pre.2024</th>\n",
       "      <th>pre.2025</th>\n",
       "      <th>post.2023</th>\n",
       "      <th>post.2024</th>\n",
       "      <th>post.2025</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Blastochloris</td>\n",
       "      <td>59282</td>\n",
       "      <td>4983.233017</td>\n",
       "      <td>-0.398737</td>\n",
       "      <td>0.121450</td>\n",
       "      <td>-3.283141</td>\n",
       "      <td>1.026572e-03</td>\n",
       "      <td>0.039921</td>\n",
       "      <td>7740</td>\n",
       "      <td>7091</td>\n",
       "      <td>7949</td>\n",
       "      <td>3194</td>\n",
       "      <td>2859</td>\n",
       "      <td>3680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Polymorphum</td>\n",
       "      <td>991903</td>\n",
       "      <td>1685.836432</td>\n",
       "      <td>-0.388073</td>\n",
       "      <td>0.098087</td>\n",
       "      <td>-3.956401</td>\n",
       "      <td>7.608749e-05</td>\n",
       "      <td>0.016623</td>\n",
       "      <td>2554</td>\n",
       "      <td>2532</td>\n",
       "      <td>2657</td>\n",
       "      <td>1056</td>\n",
       "      <td>1016</td>\n",
       "      <td>1208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alicycliphilus</td>\n",
       "      <td>201096</td>\n",
       "      <td>1367.673883</td>\n",
       "      <td>-0.387803</td>\n",
       "      <td>0.108749</td>\n",
       "      <td>-3.566037</td>\n",
       "      <td>3.624201e-04</td>\n",
       "      <td>0.028359</td>\n",
       "      <td>1877</td>\n",
       "      <td>2280</td>\n",
       "      <td>2136</td>\n",
       "      <td>813</td>\n",
       "      <td>869</td>\n",
       "      <td>981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Thermobispora</td>\n",
       "      <td>147067</td>\n",
       "      <td>1741.908430</td>\n",
       "      <td>-0.374820</td>\n",
       "      <td>0.106888</td>\n",
       "      <td>-3.506654</td>\n",
       "      <td>4.537793e-04</td>\n",
       "      <td>0.028374</td>\n",
       "      <td>2771</td>\n",
       "      <td>2687</td>\n",
       "      <td>2577</td>\n",
       "      <td>1096</td>\n",
       "      <td>1095</td>\n",
       "      <td>1178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chelatococcus</td>\n",
       "      <td>28209</td>\n",
       "      <td>4136.936104</td>\n",
       "      <td>-0.369665</td>\n",
       "      <td>0.116961</td>\n",
       "      <td>-3.160584</td>\n",
       "      <td>1.574532e-03</td>\n",
       "      <td>0.047427</td>\n",
       "      <td>6549</td>\n",
       "      <td>6053</td>\n",
       "      <td>6523</td>\n",
       "      <td>2689</td>\n",
       "      <td>2390</td>\n",
       "      <td>2917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Starkeya</td>\n",
       "      <td>152053</td>\n",
       "      <td>5597.912921</td>\n",
       "      <td>-0.357750</td>\n",
       "      <td>0.113382</td>\n",
       "      <td>-3.155261</td>\n",
       "      <td>1.603548e-03</td>\n",
       "      <td>0.047427</td>\n",
       "      <td>8845</td>\n",
       "      <td>8290</td>\n",
       "      <td>8868</td>\n",
       "      <td>3603</td>\n",
       "      <td>3217</td>\n",
       "      <td>3965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pannonibacter</td>\n",
       "      <td>227873</td>\n",
       "      <td>1326.701856</td>\n",
       "      <td>-0.357417</td>\n",
       "      <td>0.109139</td>\n",
       "      <td>-3.274871</td>\n",
       "      <td>1.057102e-03</td>\n",
       "      <td>0.039921</td>\n",
       "      <td>2084</td>\n",
       "      <td>2030</td>\n",
       "      <td>2049</td>\n",
       "      <td>852</td>\n",
       "      <td>822</td>\n",
       "      <td>876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Xanthobacter</td>\n",
       "      <td>279</td>\n",
       "      <td>2080.650012</td>\n",
       "      <td>-0.355472</td>\n",
       "      <td>0.106417</td>\n",
       "      <td>-3.340374</td>\n",
       "      <td>8.366578e-04</td>\n",
       "      <td>0.038606</td>\n",
       "      <td>3227</td>\n",
       "      <td>3148</td>\n",
       "      <td>3302</td>\n",
       "      <td>1316</td>\n",
       "      <td>1208</td>\n",
       "      <td>1484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Azorhizobium</td>\n",
       "      <td>6</td>\n",
       "      <td>1985.394327</td>\n",
       "      <td>-0.342947</td>\n",
       "      <td>0.100676</td>\n",
       "      <td>-3.406454</td>\n",
       "      <td>6.581268e-04</td>\n",
       "      <td>0.032704</td>\n",
       "      <td>3070</td>\n",
       "      <td>3070</td>\n",
       "      <td>3140</td>\n",
       "      <td>1253</td>\n",
       "      <td>1164</td>\n",
       "      <td>1391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Pontivivens</td>\n",
       "      <td>1844015</td>\n",
       "      <td>1066.529393</td>\n",
       "      <td>-0.339530</td>\n",
       "      <td>0.097943</td>\n",
       "      <td>-3.466604</td>\n",
       "      <td>5.270772e-04</td>\n",
       "      <td>0.028374</td>\n",
       "      <td>1600</td>\n",
       "      <td>1659</td>\n",
       "      <td>1736</td>\n",
       "      <td>639</td>\n",
       "      <td>650</td>\n",
       "      <td>757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Skermanella</td>\n",
       "      <td>204447</td>\n",
       "      <td>2430.713329</td>\n",
       "      <td>-0.336395</td>\n",
       "      <td>0.103174</td>\n",
       "      <td>-3.260452</td>\n",
       "      <td>1.112348e-03</td>\n",
       "      <td>0.039921</td>\n",
       "      <td>3532</td>\n",
       "      <td>3838</td>\n",
       "      <td>4035</td>\n",
       "      <td>1384</td>\n",
       "      <td>1479</td>\n",
       "      <td>1807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Rhodopseudomonas</td>\n",
       "      <td>1073</td>\n",
       "      <td>16732.103784</td>\n",
       "      <td>-0.327463</td>\n",
       "      <td>0.104012</td>\n",
       "      <td>-3.148311</td>\n",
       "      <td>1.642169e-03</td>\n",
       "      <td>0.047427</td>\n",
       "      <td>26872</td>\n",
       "      <td>25176</td>\n",
       "      <td>26586</td>\n",
       "      <td>10496</td>\n",
       "      <td>9744</td>\n",
       "      <td>11704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Roseomonas</td>\n",
       "      <td>125216</td>\n",
       "      <td>3116.125064</td>\n",
       "      <td>-0.315819</td>\n",
       "      <td>0.081321</td>\n",
       "      <td>-3.883587</td>\n",
       "      <td>1.029265e-04</td>\n",
       "      <td>0.016623</td>\n",
       "      <td>4812</td>\n",
       "      <td>4947</td>\n",
       "      <td>4963</td>\n",
       "      <td>1872</td>\n",
       "      <td>1902</td>\n",
       "      <td>2156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Salipiger</td>\n",
       "      <td>263377</td>\n",
       "      <td>2863.449134</td>\n",
       "      <td>-0.305330</td>\n",
       "      <td>0.085078</td>\n",
       "      <td>-3.588841</td>\n",
       "      <td>3.321520e-04</td>\n",
       "      <td>0.028359</td>\n",
       "      <td>4354</td>\n",
       "      <td>4529</td>\n",
       "      <td>4705</td>\n",
       "      <td>1698</td>\n",
       "      <td>1724</td>\n",
       "      <td>2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Micrococcus</td>\n",
       "      <td>1269</td>\n",
       "      <td>2011.733733</td>\n",
       "      <td>-0.301700</td>\n",
       "      <td>0.096519</td>\n",
       "      <td>-3.125808</td>\n",
       "      <td>1.773172e-03</td>\n",
       "      <td>0.047728</td>\n",
       "      <td>3162</td>\n",
       "      <td>3092</td>\n",
       "      <td>3299</td>\n",
       "      <td>1238</td>\n",
       "      <td>1225</td>\n",
       "      <td>1343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Bosea</td>\n",
       "      <td>85413</td>\n",
       "      <td>16026.625435</td>\n",
       "      <td>-0.299723</td>\n",
       "      <td>0.095448</td>\n",
       "      <td>-3.140158</td>\n",
       "      <td>1.688569e-03</td>\n",
       "      <td>0.047427</td>\n",
       "      <td>25470</td>\n",
       "      <td>24530</td>\n",
       "      <td>26155</td>\n",
       "      <td>9854</td>\n",
       "      <td>9350</td>\n",
       "      <td>11140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Defluviicoccus</td>\n",
       "      <td>111830</td>\n",
       "      <td>1210.846317</td>\n",
       "      <td>-0.299059</td>\n",
       "      <td>0.094935</td>\n",
       "      <td>-3.150157</td>\n",
       "      <td>1.631829e-03</td>\n",
       "      <td>0.047427</td>\n",
       "      <td>1829</td>\n",
       "      <td>1953</td>\n",
       "      <td>1979</td>\n",
       "      <td>690</td>\n",
       "      <td>759</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Hypericibacter</td>\n",
       "      <td>2705399</td>\n",
       "      <td>7267.970751</td>\n",
       "      <td>-0.298487</td>\n",
       "      <td>0.084239</td>\n",
       "      <td>-3.543335</td>\n",
       "      <td>3.951008e-04</td>\n",
       "      <td>0.028359</td>\n",
       "      <td>10801</td>\n",
       "      <td>11831</td>\n",
       "      <td>11962</td>\n",
       "      <td>4110</td>\n",
       "      <td>4584</td>\n",
       "      <td>5082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Methylosinus</td>\n",
       "      <td>425</td>\n",
       "      <td>3161.662910</td>\n",
       "      <td>-0.294117</td>\n",
       "      <td>0.089769</td>\n",
       "      <td>-3.276391</td>\n",
       "      <td>1.051429e-03</td>\n",
       "      <td>0.039921</td>\n",
       "      <td>4795</td>\n",
       "      <td>5076</td>\n",
       "      <td>5199</td>\n",
       "      <td>1762</td>\n",
       "      <td>2002</td>\n",
       "      <td>2223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Phreatobacter</td>\n",
       "      <td>1632780</td>\n",
       "      <td>6487.054304</td>\n",
       "      <td>-0.292036</td>\n",
       "      <td>0.084155</td>\n",
       "      <td>-3.470209</td>\n",
       "      <td>5.200538e-04</td>\n",
       "      <td>0.028374</td>\n",
       "      <td>10094</td>\n",
       "      <td>10156</td>\n",
       "      <td>10679</td>\n",
       "      <td>3915</td>\n",
       "      <td>3862</td>\n",
       "      <td>4477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Methylobacterium</td>\n",
       "      <td>407</td>\n",
       "      <td>31711.379403</td>\n",
       "      <td>-0.289801</td>\n",
       "      <td>0.076644</td>\n",
       "      <td>-3.781122</td>\n",
       "      <td>1.561230e-04</td>\n",
       "      <td>0.016809</td>\n",
       "      <td>49748</td>\n",
       "      <td>50112</td>\n",
       "      <td>51443</td>\n",
       "      <td>18969</td>\n",
       "      <td>19089</td>\n",
       "      <td>21804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Spirosoma</td>\n",
       "      <td>107</td>\n",
       "      <td>2640.764638</td>\n",
       "      <td>0.322551</td>\n",
       "      <td>0.084425</td>\n",
       "      <td>3.820574</td>\n",
       "      <td>1.331414e-04</td>\n",
       "      <td>0.016809</td>\n",
       "      <td>5019</td>\n",
       "      <td>5236</td>\n",
       "      <td>5311</td>\n",
       "      <td>1218</td>\n",
       "      <td>1341</td>\n",
       "      <td>1471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Triticum</td>\n",
       "      <td>4564</td>\n",
       "      <td>11488.251470</td>\n",
       "      <td>0.427504</td>\n",
       "      <td>0.079382</td>\n",
       "      <td>5.385414</td>\n",
       "      <td>7.227809e-08</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>22224</td>\n",
       "      <td>23624</td>\n",
       "      <td>24064</td>\n",
       "      <td>5052</td>\n",
       "      <td>5538</td>\n",
       "      <td>6249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Nicotiana</td>\n",
       "      <td>4085</td>\n",
       "      <td>1227.239461</td>\n",
       "      <td>0.515071</td>\n",
       "      <td>0.132534</td>\n",
       "      <td>3.886334</td>\n",
       "      <td>1.017695e-04</td>\n",
       "      <td>0.016623</td>\n",
       "      <td>2712</td>\n",
       "      <td>2423</td>\n",
       "      <td>2509</td>\n",
       "      <td>577</td>\n",
       "      <td>580</td>\n",
       "      <td>570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                taxa    taxid      baseMean  log2FoldChange     lfcSE  \\\n",
       "0      Blastochloris    59282   4983.233017       -0.398737  0.121450   \n",
       "1        Polymorphum   991903   1685.836432       -0.388073  0.098087   \n",
       "2     Alicycliphilus   201096   1367.673883       -0.387803  0.108749   \n",
       "3      Thermobispora   147067   1741.908430       -0.374820  0.106888   \n",
       "4      Chelatococcus    28209   4136.936104       -0.369665  0.116961   \n",
       "5           Starkeya   152053   5597.912921       -0.357750  0.113382   \n",
       "6      Pannonibacter   227873   1326.701856       -0.357417  0.109139   \n",
       "7       Xanthobacter      279   2080.650012       -0.355472  0.106417   \n",
       "8       Azorhizobium        6   1985.394327       -0.342947  0.100676   \n",
       "9        Pontivivens  1844015   1066.529393       -0.339530  0.097943   \n",
       "10       Skermanella   204447   2430.713329       -0.336395  0.103174   \n",
       "11  Rhodopseudomonas     1073  16732.103784       -0.327463  0.104012   \n",
       "12        Roseomonas   125216   3116.125064       -0.315819  0.081321   \n",
       "13         Salipiger   263377   2863.449134       -0.305330  0.085078   \n",
       "14       Micrococcus     1269   2011.733733       -0.301700  0.096519   \n",
       "15             Bosea    85413  16026.625435       -0.299723  0.095448   \n",
       "16    Defluviicoccus   111830   1210.846317       -0.299059  0.094935   \n",
       "17    Hypericibacter  2705399   7267.970751       -0.298487  0.084239   \n",
       "18      Methylosinus      425   3161.662910       -0.294117  0.089769   \n",
       "19     Phreatobacter  1632780   6487.054304       -0.292036  0.084155   \n",
       "20  Methylobacterium      407  31711.379403       -0.289801  0.076644   \n",
       "21         Spirosoma      107   2640.764638        0.322551  0.084425   \n",
       "22          Triticum     4564  11488.251470        0.427504  0.079382   \n",
       "23         Nicotiana     4085   1227.239461        0.515071  0.132534   \n",
       "\n",
       "        stat        pvalue      padj  pre.2023  pre.2024  pre.2025  post.2023  \\\n",
       "0  -3.283141  1.026572e-03  0.039921      7740      7091      7949       3194   \n",
       "1  -3.956401  7.608749e-05  0.016623      2554      2532      2657       1056   \n",
       "2  -3.566037  3.624201e-04  0.028359      1877      2280      2136        813   \n",
       "3  -3.506654  4.537793e-04  0.028374      2771      2687      2577       1096   \n",
       "4  -3.160584  1.574532e-03  0.047427      6549      6053      6523       2689   \n",
       "5  -3.155261  1.603548e-03  0.047427      8845      8290      8868       3603   \n",
       "6  -3.274871  1.057102e-03  0.039921      2084      2030      2049        852   \n",
       "7  -3.340374  8.366578e-04  0.038606      3227      3148      3302       1316   \n",
       "8  -3.406454  6.581268e-04  0.032704      3070      3070      3140       1253   \n",
       "9  -3.466604  5.270772e-04  0.028374      1600      1659      1736        639   \n",
       "10 -3.260452  1.112348e-03  0.039921      3532      3838      4035       1384   \n",
       "11 -3.148311  1.642169e-03  0.047427     26872     25176     26586      10496   \n",
       "12 -3.883587  1.029265e-04  0.016623      4812      4947      4963       1872   \n",
       "13 -3.588841  3.321520e-04  0.028359      4354      4529      4705       1698   \n",
       "14 -3.125808  1.773172e-03  0.047728      3162      3092      3299       1238   \n",
       "15 -3.140158  1.688569e-03  0.047427     25470     24530     26155       9854   \n",
       "16 -3.150157  1.631829e-03  0.047427      1829      1953      1979        690   \n",
       "17 -3.543335  3.951008e-04  0.028359     10801     11831     11962       4110   \n",
       "18 -3.276391  1.051429e-03  0.039921      4795      5076      5199       1762   \n",
       "19 -3.470209  5.200538e-04  0.028374     10094     10156     10679       3915   \n",
       "20 -3.781122  1.561230e-04  0.016809     49748     50112     51443      18969   \n",
       "21  3.820574  1.331414e-04  0.016809      5019      5236      5311       1218   \n",
       "22  5.385414  7.227809e-08  0.000047     22224     23624     24064       5052   \n",
       "23  3.886334  1.017695e-04  0.016623      2712      2423      2509        577   \n",
       "\n",
       "    post.2024  post.2025  \n",
       "0        2859       3680  \n",
       "1        1016       1208  \n",
       "2         869        981  \n",
       "3        1095       1178  \n",
       "4        2390       2917  \n",
       "5        3217       3965  \n",
       "6         822        876  \n",
       "7        1208       1484  \n",
       "8        1164       1391  \n",
       "9         650        757  \n",
       "10       1479       1807  \n",
       "11       9744      11704  \n",
       "12       1902       2156  \n",
       "13       1724       2014  \n",
       "14       1225       1343  \n",
       "15       9350      11140  \n",
       "16        759        846  \n",
       "17       4584       5082  \n",
       "18       2002       2223  \n",
       "19       3862       4477  \n",
       "20      19089      21804  \n",
       "21       1341       1471  \n",
       "22       5538       6249  \n",
       "23        580        570  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elite_deseq=pd.read_csv('deseq2/elite.txt',sep=\"\\t\")\n",
    "elite_deseq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d3e278-f654-43b2-8a47-711f9300aec9",
   "metadata": {},
   "source": [
    "Elite shows reductions in a small number of baterial genera, all of which have decent numbers of counts associated with them. Interestingly, Triticum and Nicotiana and increased, presumably as a result of the removal of a considerable amount of sequences effectively enriching them. The Triticum in question is emmer wheat. \n",
    "\n",
    "We'll just focus on elite for now, since this is where we seem to see a problem.\n",
    "\n",
    "### Collect taxids...\n",
    "\n",
    "We have taxids at the genus level for the species of interest, but some classifications may be at a lower level, so we need to get the full set of taxids. The output of kraken-inspect which is included with the database contains all the details of the included taxa, so we can trawl this to find ids of relevance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "59b2826f-c46c-4f22-8dbb-fa86390bb67c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genus_ids=elite_deseq['taxid'].tolist()\n",
    "genus_ids=list(map(str, genus_ids))\n",
    "tax_ids=dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aa38da7-6d22-4d58-8a7a-2bcaea476d6a",
   "metadata": {},
   "source": [
    "The list of genus taxids should be screened to select only the bacteria. Let's reuse our existing taxonomy database code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d86c7fe5-00f0-4c07-baed-1e56f411fee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('bin') \n",
    "from taxonomy import download_taxdb,parse_taxonomy\n",
    "\n",
    "if not os.path.isdir('taxonomy'):\n",
    "    os.mkdir('taxonomy')\n",
    "    download_taxdb('taxonomy')\n",
    "    \n",
    "nodes,names=parse_taxonomy('taxonomy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f82d0334-b426-4fb1-bbe0-555ffcf41d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_bacteria(taxid,nodes):\n",
    "    if int(taxid) in nodes:\n",
    "        node_data=nodes[int(taxid)]\n",
    "        rank=node_data['rank']\n",
    "        while node_data['rank'] != 'superkingdom':\n",
    "            node_data=nodes[node_data['parent']]\n",
    "        if node_data['name']=='Bacteria':\n",
    "            return(True)\n",
    "        else:\n",
    "            return(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1d8b821-d25d-4a26-aa3f-911b0c9da979",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacteria_ids=list()\n",
    "for id in genus_ids:\n",
    "    if is_bacteria(id,nodes)==True:\n",
    "        bacteria_ids.append(id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c82e01-c4d9-4a91-8004-3e26dc7b36a3",
   "metadata": {},
   "source": [
    "The kraken-inspect output is a tab-delimited file with the taxonomy ids in column 5, while column 4 includes a coded taxon level i.e G, S, S2:\n",
    "\n",
    "```\n",
    "3.53  638345314       7784546 F5      163735             Phaseoleae\n",
    "1.41  254736019       16589388        G       3913         Vigna\n",
    "0.49  88769661        88769661        S       3917           Vigna unguiculata\n",
    "0.42  75549874        0       S       157791                 Vigna radiata\n",
    "0.42  75549874        75549874        S1      3916           Vigna radiata var. radiata\n",
    "0.41  73827096        73827096        S       3914           Vigna angularis\n",
    "1.02  184902669       0       G       3846                 Glycine\n",
    "  ```\n",
    "  \n",
    "All the species/subspecies ranks immediately follow the genus, so we can scan until we find the taxid of interest then consume rows with an 'S?' code in column 4 to gain the additional tax ids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3d048e23-24dd-478b-9128-f84f4e9ca8aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_ids(taxid):\n",
    "    ids=list()\n",
    "    with open('../training_data/kraken2_pluspfp_20210127/inspect.txt', 'r') as handle:\n",
    "        for line in handle:\n",
    "            # skip comments\n",
    "            if re.search('^#',line):\n",
    "                continue\n",
    "                \n",
    "            fields=line.strip().split(\"\\t\")\n",
    "            if fields[4]==taxid:\n",
    "                print(\"Found \",fields[5].lstrip())\n",
    "                ids.append(taxid)\n",
    "                genus=fields[5].lstrip()\n",
    "                for line in handle:\n",
    "                    fields=line.strip().split(\"\\t\")\n",
    "                    if re.search('^(S|G[0-9])',fields[3]):\n",
    "                        print('Found a species: ',fields[5].lstrip())\n",
    "                        ids.append(fields[4])\n",
    "                    else:\n",
    "                        return(genus,ids)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a36291cc-9dc1-4141-b0c8-f3f3d198ea38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found  Blastochloris\n",
      "Found a species:  Blastochloris tepida\n",
      "Found a species:  Blastochloris viridis\n",
      "Found  Polymorphum\n",
      "Found a species:  Polymorphum gilvum\n",
      "Found a species:  Polymorphum gilvum SL003B-26A1\n",
      "Found  Alicycliphilus\n",
      "Found a species:  Alicycliphilus denitrificans\n",
      "Found a species:  Alicycliphilus denitrificans K601\n",
      "Found a species:  Alicycliphilus denitrificans BC\n",
      "Found  Thermobispora\n",
      "Found a species:  Thermobispora bispora\n",
      "Found a species:  Thermobispora bispora DSM 43833\n",
      "Found  Chelatococcus\n",
      "Found a species:  unclassified Chelatococcus\n",
      "Found a species:  Chelatococcus sp. CO-6\n",
      "Found a species:  Chelatococcus daeguensis\n",
      "Found  Starkeya\n",
      "Found a species:  unclassified Starkeya\n",
      "Found a species:  Starkeya sp. ORNL1\n",
      "Found a species:  Starkeya novella\n",
      "Found a species:  Starkeya novella DSM 506\n",
      "Found  Pannonibacter\n",
      "Found a species:  Pannonibacter phragmitetus\n",
      "Found a species:  Pannonibacter phragmitetus BB\n",
      "Found  Xanthobacter\n",
      "Found a species:  Xanthobacter autotrophicus\n",
      "Found a species:  Xanthobacter autotrophicus Py2\n",
      "Found a species:  Xanthobacter flavus\n",
      "Found  Azorhizobium\n",
      "Found a species:  Azorhizobium caulinodans\n",
      "Found a species:  Azorhizobium caulinodans ORS 571\n",
      "Found  Pontivivens\n",
      "Found a species:  unclassified Pontivivens\n",
      "Found a species:  Pontivivens sp. MT2928\n",
      "Found  Skermanella\n",
      "Found a species:  Skermanella pratensis\n",
      "Found  Rhodopseudomonas\n",
      "Found a species:  Rhodopseudomonas palustris\n",
      "Found a species:  Rhodopseudomonas palustris BisA53\n",
      "Found a species:  Rhodopseudomonas palustris BisB18\n",
      "Found a species:  Rhodopseudomonas palustris HaA2\n",
      "Found a species:  Rhodopseudomonas palustris TIE-1\n",
      "Found a species:  Rhodopseudomonas palustris DX-1\n",
      "Found a species:  Rhodopseudomonas palustris CGA009\n",
      "Found  Roseomonas\n",
      "Found a species:  Roseomonas gilardii\n",
      "Found a species:  Roseomonas mucosa\n",
      "Found a species:  unclassified Roseomonas\n",
      "Found a species:  Roseomonas sp. FDAARGOS_362\n",
      "Found  Salipiger\n",
      "Found a species:  unclassified Salipiger\n",
      "Found a species:  Yangia sp. CCB-MM3\n",
      "Found a species:  Salipiger profundus\n",
      "Found a species:  Salipiger pacificus\n",
      "Found  Micrococcus\n",
      "Found a species:  Micrococcus luteus\n",
      "Found a species:  Micrococcus luteus NCTC 2665\n",
      "Found a species:  unclassified Micrococcus\n",
      "Found a species:  Micrococcus sp. KBS0714\n",
      "Found a species:  Micrococcus sp. V7\n",
      "Found a species:  Micrococcus sp. A1\n",
      "Found a species:  Micrococcus sp. 28\n",
      "Found a species:  Micrococcus sp. A7\n",
      "Found a species:  Micrococcus sp. MG-2010-D12\n",
      "Found  Bosea\n",
      "Found a species:  unclassified Bosea\n",
      "Found a species:  Bosea sp. F3-2\n",
      "Found a species:  Bosea sp. Tri-49\n",
      "Found a species:  Bosea sp. AS-1\n",
      "Found a species:  Bosea sp. PAMC 26642\n",
      "Found a species:  Bosea sp. RAC05\n",
      "Found a species:  Bosea sp. ANAM02\n",
      "Found a species:  Bosea vaviloviae\n",
      "Found  Defluviicoccus\n",
      "Found a species:  Defluviicoccus vanus\n",
      "Found  Hypericibacter\n",
      "Found a species:  Hypericibacter terrae\n",
      "Found a species:  Hypericibacter adhaerens\n",
      "Found  Methylosinus\n",
      "Found a species:  Methylosinus trichosporium\n",
      "Found a species:  Methylosinus trichosporium OB3b\n",
      "Found a species:  unclassified Methylosinus\n",
      "Found a species:  Methylosinus sp. C49\n",
      "Found  Phreatobacter\n",
      "Found a species:  Phreatobacter stygius\n",
      "Found a species:  unclassified Phreatobacter\n",
      "Found a species:  Phreatobacter sp. NMCR1094\n",
      "Found a species:  Phreatobacter cathodiphilus\n",
      "Found  Methylobacterium\n",
      "Found a species:  unclassified Methylobacterium\n",
      "Found a species:  Methylobacterium sp. 4-46\n",
      "Found a species:  Methylobacterium sp. WL1\n",
      "Found a species:  Methylobacterium sp. 17Sr1-43\n",
      "Found a species:  Methylobacterium sp. 17Sr1-1\n",
      "Found a species:  Methylobacterium sp. AMS5\n",
      "Found a species:  Methylobacterium sp. DM1\n",
      "Found a species:  Methylobacterium sp. C1\n",
      "Found a species:  Methylobacterium sp. XJLW\n",
      "Found a species:  Methylobacterium sp. NI91\n",
      "Found a species:  Methylobacterium sp. CLZ\n",
      "Found a species:  Methylobacterium nodulans\n",
      "Found a species:  Methylobacterium nodulans ORS 2060\n",
      "Found a species:  Methylobacterium durans\n",
      "Found a species:  Methylobacterium aquaticum\n",
      "Found a species:  Methylobacterium mesophilicum\n",
      "Found a species:  Methylobacterium mesophilicum SR1.6/6\n",
      "Found a species:  Methylobacterium currus\n",
      "Found a species:  Methylobacterium brachiatum\n",
      "Found a species:  Methylobacterium terrae\n",
      "Found a species:  Methylobacterium radiotolerans\n",
      "Found a species:  Methylobacterium radiotolerans JCM 2831\n",
      "Found a species:  Methylobacterium phyllosphaerae\n",
      "Found a species:  Methylobacterium oryzae\n",
      "Found a species:  Methylobacterium oryzae CBMB20\n",
      "Found  Spirosoma\n",
      "Found a species:  unclassified Spirosoma\n",
      "Found a species:  Spirosoma sp. PL0136\n",
      "Found a species:  Spirosoma sp. KCTC 42546\n",
      "Found a species:  Spirosoma sp. I-24\n",
      "Found a species:  Spirosoma sp. BT328\n",
      "Found a species:  Spirosoma sp. CJU-R4\n",
      "Found a species:  Spirosoma sp. KUDC1026\n",
      "Found a species:  Spirosoma sp. TS118\n",
      "Found a species:  Spirosoma pollinicola\n",
      "Found a species:  Spirosoma aerolatum\n",
      "Found a species:  Spirosoma radiotolerans\n",
      "Found a species:  Spirosoma rigui\n",
      "Found a species:  Spirosoma montaniterrae\n"
     ]
    }
   ],
   "source": [
    "for id in bacteria_ids:\n",
    "    (genus,ids)=find_ids(id)\n",
    "    tax_ids[genus]=ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e58bd5a-5a91-4789-b910-188dd42beb3d",
   "metadata": {},
   "source": [
    "### Subset Reads\n",
    "\n",
    "Now we know which tax ids we are looking for we can use the Kraken classifications to separate out the reads identified as belonging to each species of interest. This will require parsing the kraken output files to pull out the read ids which are assigned to the required tax ids, then parsing the fastq files to separate these out.\n",
    "\n",
    "Do not expect this to be quick...or pretty..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a34bbbfe-5577-41a2-89ea-209c5ee0c7cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing 2023 kraken output...\n",
      "Reading 2023 F reads\n",
      "Reading 2023 R reads\n",
      "Parsing 2024 kraken output...\n",
      "Reading 2024 F reads\n",
      "Reading 2024 R reads\n",
      "Parsing 2025 kraken output...\n",
      "Reading 2025 F reads\n",
      "Reading 2025 R reads\n"
     ]
    }
   ],
   "source": [
    "taxid_re=re.compile('taxid ([0-9]+)')\n",
    "\n",
    "if not os.path.isdir('per_genus_fastq'):\n",
    "    os.mkdir('per_genus_fastq')\n",
    "    \n",
    "for sample in samples['elite']:\n",
    "    read_ids={} # dict to store read ids per genus\n",
    "    print('Parsing {} kraken output...'.format(sample))\n",
    "    with open('kraken/nt_unfiltered/{}.txt'.format(sample)) as handle:\n",
    "        for line in handle:\n",
    "            fields=line.strip().split(\"\\t\")\n",
    "            m=re.search(taxid_re,fields[2])\n",
    "            if m:\n",
    "                read_tax_id=m.group(1)\n",
    "                for genus in tax_ids.keys():\n",
    "                    if read_tax_id in tax_ids[genus]:\n",
    "                        if genus in read_ids.keys():\n",
    "                            read_ids[genus].append(fields[1])                            \n",
    "                        else:\n",
    "                            read_ids[genus]=[fields[1]]        \n",
    "\n",
    "    # uncompress reads to $TMPDIR and index for random access...\n",
    "    for read in ('1','2'):\n",
    "        with gzip.open('fastq/{}_{}.fq.gz'.format(sample,read),'rb') as f_in:\n",
    "            with open('{}/{}_{}.fq'.format(os.environ['TMPDIR'],sample,read), 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "                \n",
    "    print('Reading {} F reads'.format(sample))\n",
    "    f_reads = SeqIO.index('{}/{}_1.fq'.format(os.environ['TMPDIR'],sample), \"fastq\")\n",
    "    print('Reading {} R reads'.format(sample))\n",
    "    r_reads = SeqIO.index('{}/{}_2.fq'.format(os.environ['TMPDIR'],sample), \"fastq\")\n",
    "\n",
    "    # Collect reads for each genus and write to separate fastq files...\n",
    "    for genus in read_ids.keys():   \n",
    "        genus_f_reads=[]\n",
    "        genus_r_reads=[]\n",
    "        \n",
    "        for read in read_ids[genus]:\n",
    "            try:\n",
    "                f=f_reads[read]\n",
    "            except KeyError:\n",
    "                print('{} f read not found'.format(read))\n",
    "            \n",
    "            try:\n",
    "                r=r_reads[read]\n",
    "            except KeyError:\n",
    "                print('{} r read not found'.format(read))\n",
    "            \n",
    "            if f and r:\n",
    "                genus_f_reads.append(f)\n",
    "                genus_r_reads.append(r)\n",
    "    \n",
    "        with open('per_genus_fastq/{}_1.fq'.format(genus),'a') as out:\n",
    "            SeqIO.write(genus_f_reads, out, 'fastq')\n",
    "        with open('per_genus_fastq/{}_2.fq'.format(genus),'a') as out:\n",
    "            SeqIO.write(genus_r_reads, out, 'fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "978ec6d7-3ce0-48c2-8754-6daa59051205",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now compress the finished per-genus fastq files...\n",
    "for genus in read_ids.keys():\n",
    "    for read in ('1','2'):\n",
    "        with open('per_genus_fastq/{}_{}.fq'.format(genus,read), 'rb') as f_in:\n",
    "            with gzip.open('per_genus_fastq/{}_{}.fq.gz'.format(genus,read), 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "        os.remove('per_genus_fastq/{}_{}.fq'.format(genus,read))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "623daef6-497a-4a2e-8fb8-acd4278baf13",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:miniconda3-host_filter]",
   "language": "python",
   "name": "conda-env-miniconda3-host_filter-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
